---
title: "Create Phyloseq object"
author: "Mari-Lee Odendaal"
date: "`r Sys.time()`"
---

The overall goal of this script is to merge phyloseq objects generated by our DADA2 pipeline from each run and roughly filter the crude phyloseq object. This phyloseq is properly cleaned and filtered in a later script. Alongside the phyloseq object, we closely note the details of each filtering step.

```{r setup, message=F, include=F}
subdir_name <- "1_create_phyloseq"

library(tidyverse);library(magrittr);library(glue);library(here);library(fs);library(phyloseq);library(microbiomer);library(microbiome);library(reshape2);library(scales);library(dada2);library(janitor);library("xlsx");library(lubridate);library(ggtext);library(ggnewscale);library(vegan);library(gtsummary)

# set paths
knitr::opts_knit$set(root.dir=".", aliases=c(h = "fig.height", w = "fig.width", ow = "out.width"))
knitr::opts_chunk$set(dev=c('png', 'pdf'), 
                      fig.path=here("results", "figures", glue("{subdir_name}/")),
                      dpi=300)

theme_set(theme_bw())
theme_update(axis.text.x = element_text(angle = 45, hjust=1),
             strip.text = element_text(colour = 'white'),
             strip.background =element_rect(fill="azure4", color = "azure4"))
```

```{r knit, echo=F, eval=FALSE, warning = F, include=F}
rmarkdown::render(input = here("scripts", str_c(subdir_name, ".Rmd")), output_dir = here("results"))
```

# Set-up
## Functions
The functions used in this RMarkdown file are stored in "create_phyloseq.R". This file includes functions to create, clean and order the phyloseq object.

```{r functions}
ps_stats <- function(ps, stats, step_name) {
  stats <- add_row(stats, step = step_name, 
                   samples = nsamples(ps), 
                   taxa = ntaxa(ps), 
                   reads = sum(otu_table(ps)))
  return(stats)
}

create_taxa_filter <- function(taxa) {
  #remove Chloroplast/Mitochondria/Archae/Eukaryota/no kingdom-level annotation
  #see Park et al. (Raes). Microbiome. 2019. / Winkler et al. Cell. 2020.
  taxa_filter <- (str_detect(replace_na(taxa[, "Family"], ""), "Mitochondria") |
                    str_detect(replace_na(taxa[, "Order"], ""), "Chloroplast") |
                    str_detect(replace_na(taxa[, "Kingdom"], ""), "Archae|Eukaryota") |
                    is.na(taxa[, "Kingdom"]))
  return(taxa_filter)
}

create_taxa_names <- function(tax_table) {
  tax_names <- apply(tax_table, 1, function(x) {
    if(any(names(na.omit(x)) %in% "ASV")) {
      name <- str_c(tail(na.omit(x), 2), collapse = "_")
    } else { 
      name <- tail(na.omit(x), 1)
    }
    return(name)
  })
  
  tax_names %<>% 
    str_replace_all("-| ", "_") %>%
    str_remove_all("\\[|\\]|\\(|\\)")
  
  return(str_c(tax_names, 1:nrow(tax_table), sep = "_"))
}

create_phylo_manual <- function(seqtab, taxa, meta) {
  ps <- phyloseq(otu_table(t(seqtab), taxa_are_rows=T), 
                 sample_data(meta), # TODO add meta
                 tax_table(taxa))
  dna <- Biostrings::DNAStringSet(taxa_names(ps))
  names(dna) <- taxa_names(ps)
  ps <- merge_phyloseq(ps, dna)
  taxa_names(ps) <- create_taxa_names(taxa)
  tax_table(ps) <- cbind(tax_table(ps), ASV = taxa_names(ps))
  return(ps)
}

create_ps <- function(seqtab, taxa, meta) {
  
  #order mapping/seqtab
  seqtab_ordered <- seqtab[match(rownames(meta), rownames(seqtab)), ]
  
  #filter taxa
  taxa_filter <- create_taxa_filter(taxa)
  
  taxa_filtered <- taxa[!taxa_filter, ]
  seqtab_filtered <- seqtab_ordered[, !taxa_filter]
  n_taxa_excl <- nrow(taxa) - nrow(taxa_filtered)
  
  print(glue::glue("ASVs beloning to the phylogenetic groups 'Mitochondria', 'Chloroplast', 'Archae' or 'Eukaryota' were
                   excluded (n = {n_taxa_excl} ASVs), resulting in a total of {scales::comma(nrow(taxa))} ASVs after exclusion."))
  
  ps <- create_phylo_manual(seqtab_filtered, taxa_filtered, meta)
  
  return(ps)
}

create_crude_ps <- function(seqtab, species) {
  ps <- phyloseq(otu_table(t(seqtab), taxa_are_rows = T), tax_table(species %>% as.matrix))
  return(ps)
}

remove_c7_taxa <- function(all_ps) {
  x <- vector("list")
  for(i in all_ps) {
    tax_table(i) <- tax_table(i)[,c(1,2,3,4,5,6,8)]
    x[[length(x) + 1]] <- i}
  return(x)}

merge_ps_crude <- function(dada2_files) {
  all_ps <- dada2_files$ps_crude
  names(all_ps) <- dada2_files$name
  list2env(all_ps, envir = environment())
  ps_merged <-  do.call(merge_phyloseq, map(names(all_ps), as.symbol))
  return(ps_merged)
}

taxa_sum_reorder <- function(ps) {
  index <- order(taxa_sums(ps), decreasing = T)
  tax_table(ps) <- tax_table(ps)[index,]
  otu_table(ps) <- otu_table(ps)[,index]
  return(ps)
}

ps_stats <- function(ps, stats, step_name) {
  stats <- add_row(stats, step = step_name, 
                   samples = nsamples(ps), 
                   taxa = ntaxa(ps), 
                   reads = sum(otu_table(ps)))
  return(stats)
}

separate_species <- function(ps) {
  seqtab <- as(otu_table(ps), "matrix")
  taxa <- as.data.frame(tax_table(ps)) %>% separate(Species, c('species1', 'species2', "species3"), sep = "/", remove = F) %>% 
    mutate(Species_one = ifelse(is.na(species2), species1, NA),
           Species_one = ifelse(is.na(Genus), NA, Species_one),
           Species_one = ifelse(is.na(Species_one), Genus, paste0(Genus, "_", Species_one)))
  taxa2 <- subset(taxa, select = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Species_one")) %>% as.matrix()
  
  taxa2[taxa2 == "Prevotella_7"] <- "Prevotella"
  
  meta2 <- sample_data(ps) %>% as.data.frame()
  ps <- create_ps(seqtab, taxa2, meta2) 

  return(ps)
}
```

# Pre-processing
## Merging phyloseq objects of each MiSeq run 
We have used the RIVM DADA2 pipeline to process all the sequence runs into an understandable count table format. Here we load each of the phyloseq objects (n = 11) generated by the pipeline into the R environment. 

```{r}
ps54 <- readRDS(file = here("input", "L054_VGO3_phyloseq.rds"))
sample_names(ps54) <- gsub("_", "-", sample_names(ps54))
ps56 <- readRDS(file = here("input", "L056_VGO3_phyloseq.rds"))
ps57 <- readRDS(file = here("input", "L057_VGO3_phyloseq.rds"))
ps58 <- readRDS(file = here("input", "L058_VGO3_phyloseq.rds"))
ps59 <- readRDS(file = here("input", "L059_VGO3_phyloseq.rds"))
ps60 <- readRDS(file = here("input", "L060_VGO3_phyloseq.rds"))
ps61 <- readRDS(file = here("input", "L061_VGO3_phyloseq.rds"))
ps62 <- readRDS(file = here("input", "L062_VGO3_phyloseq.rds"))
ps63 <- readRDS(file = here("input", "L063_VGO3_phyloseq.rds"))
ps64 <- readRDS(file = here("input", "L064_VGO3_phyloseq.rds"))
ps65 <- readRDS(file = here("input", "L065_VGO3_phyloseq.rds"))

meta <- read.csv(here("input", "VGO3_labandparticipant_metadata.csv"), dec = ".", sep = ",") %>% 
  mutate(across(c(ct.value, X16S.QPCR.pg.ul, Picogreen.ug.ul), as.numeric))
meta$assay_sample <- gsub("_", "-", meta$assay_sample)

ps1 <- merge_phyloseq(ps54,ps56,ps57,ps58,ps59,ps60,ps61,ps62,ps63,ps64,ps65)
meta[!meta$assay_sample %in% sample_names(ps1),]
ps1
```

We remove the "samples" that have less than 100 reads. These "samples" are likely artefacts generated by a mismatch in either one of the barcode pairs. 

```{r}
ps1@sam_data$read_nr <- rowSums(ps1@otu_table)
ps1 <- prune_samples(ps1@sam_data$read_nr > 100, ps1)
ps1 <- prune_samples(sample_names(ps1) != "R1-001", ps1)
ps1 <- prune_samples(sample_names(ps1) != "R1_001", ps1)
ps1
```

```{r}
meta <- meta[meta$sample_id != "0",]
sample_names(ps1) %in% meta$assay_sample %>% table()
#sample_names(ps1)[!sample_names(ps1) %in% meta$assay_sample]
#meta$assay_sample[!meta$assay_sample %in% sample_names(ps1)]

m <- meta_to_df(ps1)
m$assay_sample <- m$sample_id
meta[!meta$assay_sample %in% m$assay_sample,]
m <- m %>% select(-c(X, sample_identifier, Filtered, denoised_R1, denoised_R2,
                     Merged, non.chim, non.chim...filtered.by.length, Sample))
colnames(m) <- c("sample_id", "input_reads_nr", "perc_readds_kept_after_DADA2", "after_DADA2_reads_nr", "assay_sample")

m <- left_join(m, meta, by = "assay_sample")
m <- m %>% rename(sample_id = sample_id.x) %>% select(-sample_id.y)
colnames(m) <- gsub("X16S.QPCR.pg.ul", "qPCR", colnames(m))
m$DI.Run <- as.character(m$DI.Run)
m <- subset(m, niche != "?NP")
m$niche <- factor(m$niche, levels = c("NP", "OP", "BD", "BP", "FB", 
                                      "ZMCB", "ZMCD", "UMD", "BB"))
m$DI.Run <- factor(m$DI.Run, levels = m$DI.Run %>% unique() %>% as.numeric() %>% sort() %>% as.character())
```

We use the Zoo package to convert the date variables to actual dates.

```{r}
library(zoo)
m$visit_date <- as.Date(m$visit_date, format = "%Y-%m-%d")
m$DI.Date <- as.Date(m$DI.Date, origin = "1900-01-01")
m$pneumonia_lastyr_date <- as.Date(m$pneumonia_lastyr_date, format = "%Y-%m-%d")
m$acute_resp_infec_lastyr_date <- as.Date(m$acute_resp_infec_lastyr_date, format = "%Y-%m-%d")
m$covid19_confirmed_date <- as.Date(m$covid19_confirmed_date, format = "%Y-%m-%d")
m$covid19_suspected_date <- as.Date(m$covid19_suspected_date, format = "%Y-%m-%d")
m$covid19_vaccine1_date <- as.Date(m$covid19_vaccine1_date, format = "%Y-%m-%d")
m$covid19_vaccine2_date <- as.Date(m$covid19_vaccine2_date, format = "%Y-%m-%d")
m$antibiotics_lastyr_date <- as.Date(m$antibiotics_lastyr_date, format = "%Y-%m-%d")
m$antacids_lastyr_date <- as.Date(m$antacids_lastyr_date, format = "%Y-%m-%d")
m$covid19_vaccine3_date <- as.Date(m$covid19_vaccine3_date, format = "%Y-%m-%d")
m$fuquestdate <- as.Date(m$fuquestdate, format = "%Y-%m-%d")
m$pneumonia_sympt_onset_date <- as.Date(m$pneumonia_sympt_onset_date, format = "%Y-%m-%d")
m$lamb_oldest_date <- as.Date(m$lamb_oldest_date, format = "%Y-%m-%d")
m$lamb_youngest_date <- as.Date(m$lamb_youngest_date, format = "%Y-%m-%d")
#m$special_work_date <- as.Date(m$special_work_date, format = "%Y-%m-%d")
#m$questionnaire_date <- as.Date(m$questionnaire_date, format = "%Y-%m-%d")
```

The total number of samples per population is as follows:

```{r}
#table(m$population, m$Miseq.Run, m$niche)
table(m$niche, m$population)
```

We also add a column for sampling_season and sampling_month using the visit_date information.

```{r}
m <- m %>% mutate(date1 =  as.Date(format(visit_date, '2021-%m-%d')), 
                  sampling_season = case_when(between(date1, ymd('2021-03-20'),
                                                      ymd('2021-06-19'))~ 'Spring',
                                              between(date1, ymd('2021-06-20'), 
                                                      ymd('2021-09-22'))~ 'Summer',
                                              between(date1, ymd('2021-09-22'), 
                                                      ymd('2021-12-21')) ~ 'Fall',
                                              TRUE ~ 'Winter'), date1 = NULL,
                  sampling_month = month(visit_date))
```

# Descriptive statistics
Using the data we have so far, we conduct simple descriptive statistics

```{r descriptive_statistics, ow = '90%', h = 4, w = 5}
m_sub <- subset(m, niche != "FB")
m_sub <- subset(m_sub, population != "LC")
m_sub <- subset(m_sub, niche != "NP")

m_sub <- tibble("Age in years" = m_sub$age,
                "Gender" = m_sub$gender,
                "Population" = m_sub$population,
                "Season" = m_sub$sampling_season)

m_sub$Population <- factor(m_sub$Population, levels = c("CP", "GP", "GF"))

tb <- m_sub %>%
  tbl_summary(by = "Population",
              type = c("Gender", "Season") ~ "categorical",
              statistic = list(all_continuous() ~ "{mean} ({min}, {max})")) %>% 
  modify_header(label = "**Characteristics**") %>% 
  bold_labels() 
tb
```

```{r}
export_to_excel <- function(table) {
  table %>%
    gtsummary::as_tibble() %>% 
    rename_with(~str_remove_all(., "\\**"))
}  

dir_path = here::here("results", "tables")
write.xlsx(export_to_excel(tb), glue::glue("{dir_path}/baseline.xlsx"), 
           sheetName="baseline", row.names=T)
```

Using Chi2 tests we found that season of sampling was not balanced between the three populations: 0 goat farmer samples were collected in spring.

```{r}
chisq.test(m_sub$Population, m_sub$Gender)
chisq.test(m_sub$Population, m_sub$Season)
aov(m_sub$`Age in years` ~ m_sub$Population) %>% summary()
```

We create a visual overview of the samples across the three populations:

```{r sample_distribution, ow = '90%', h = 4, w = 5}
t <- count(m, population, niche) %>% as.data.frame()
t <- subset(t, !is.na(population))

ggplot(t, aes(x = population, y = n, fill = niche)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("burlywood3", "cadetblue3", "coral3", "coral4", 
                               "darkolivegreen3", "darkolivegreen4", "#FFC857", 
                               "azure3")) +
  labs(x = "Population", y = "Number of samples", fill = "Sample type") +
  geom_text(position = position_stack(vjust = 0.5), 
            aes(label = n, colour = niche), size = 3) +
  scale_colour_manual(values = c("burlywood4", "cadetblue4", "coral4", "coral2", 
                                 "darkolivegreen4", "darkolivegreen2", "#7F642B", 
                                 "azure4")) +
  guides(colour = "none")
```

# Bacterial density & Picogreen

Calculate the alpha diversity indices.

```{r}
m2 <- subset(m, niche != "NP?")
m2 <- subset(m2, niche != "BB")
```

We use the following function to visualize the qPCR and picogreen across Miseq run

```{r}
plot_sample_data <- function(df, x, y) {
  df %>%
     ggplot(aes(x = x, y = y, color=niche)) +
  geom_point(position=position_jitterdodge(), size = 2.5, alpha = 0.6, shape = 16) + 
  labs(color = "Sample type") +
  geom_boxplot(alpha = 0.4, outlier.shape = NA, fill = "white", show.legend = FALSE) + 
  theme(legend.position="top") + 
    scale_color_manual(values = c("burlywood3", "cadetblue3", "coral3", "coral4", 
    "darkolivegreen3", "darkolivegreen4", "#FFC857", "azure3")) + 
    guides(color = guide_legend(override.aes = list(alpha=1, size = 4, shape = 15), nrow = 1))
}
```

Plots for qPCR

```{r qPCR_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$qPCR) + labs(x = "Sample Type", y = "qPCR (pg/µl)") + scale_y_continuous(trans='log10',
                     breaks=trans_breaks('log10', function(x) 10^x),
                     labels=trans_format('log10', math_format(10^.x))) +
    geom_hline(aes(yintercept = 0.095),  colour="grey50", linetype="dotted") + facet_grid(population~Miseq.Run, scale = "fixed")
```

Plots for Picogreen

```{r Picogreen_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$Picogreen.ug.ul) + labs(x = "Sample Type", y = "Picogreen") + scale_y_continuous(trans='log10',
                     breaks=trans_breaks('log10', function(x) 10^x),
                     labels=trans_format('log10', math_format(10^.x))) + 
  facet_grid(population~Miseq.Run, scale = "fixed")
```

```{r reads_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$after_DADA2_reads_nr) + labs(x = "Sample Type", y = "Number of reads") +
    geom_hline(aes(yintercept = 10000),  colour="grey50", linetype="dotted") + facet_grid(population~Miseq.Run, scale = "fixed")
```

```{r reads_kept_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$perc_readds_kept_after_DADA2) + labs(x = "Sample Type", y = "% of reads kept") +
    geom_hline(aes(yintercept = 50),  colour="grey50", linetype="dotted") + facet_grid(population~Miseq.Run, scale = "fixed")
```

Filter and order the taxa

```{r}
ps1 <- prune_taxa(taxa_sums(ps1) > 0, ps1)
stats <- tibble(step = character(),
                samples = numeric(),
                taxa = numeric(),
                reads = numeric())
stats <- ps_stats(ps1, stats, "Merged phyloseq object")
```

```{r}
ps1 <- taxa_sum_reorder(ps1) %>% separate_species()
taxa_names(ps1) <- gsub("_7_", "_", taxa_names(ps1))
```

```{r}
ps1 <- prune_taxa(taxa_sums(ps1) > 0, ps1)
stats <- ps_stats(ps1, stats, "Exclusion of Mitochondria, Chloroplast, Archaea & Eukaryota")
```

Add sample data to phyloseq object

```{r}
rownames(m) <- m$sample_id
m <- m %>% select(-c(assay_sample, sample_id))
sample_data(ps1) <- m
```

Finally, we save two phyloseq object:

```{r}
dir_path = here::here("results", "RData")
saveRDS(ps1, file = glue::glue("{dir_path}/ps_raw.Rds"))
```

# Sessioninfo
The session info is as follows:

```{r}
sessionInfo()
```


