---
title: "Create Phyloseq object"
author: "Mari-Lee Odendaal (adapted by Beatrice Cornu Hewitt)"
date: "`r Sys.time()`"
output: html_document
---

In this script phyloseq objects generated by the DADA2 pipeline from each run are combined into a phyloseq object which is then filtered (crudely). Further filtering steps are performed later on in another script. 

# Packages
```{r}
# List of required packages
required_packages <- c(
  "tidyverse", "magrittr", "glue", "fs", "phyloseq", "microbiomer", 
  "microbiome", "reshape2", "scales", "janitor", "lubridate", 
  "ggtext", "ggnewscale", "vegan", "gtsummary", "openxlsx", "dplyr"
)
# Install missing packages
install_if_missing(required_packages)
# Load all required packages
lapply(required_packages, library, character.only = TRUE)
```

# Functions
The functions used in this RMarkdown file are stored in "functions.R"
```{r}
source("functions.R")
```

# Pre-processing
## Merging phyloseq objects of each MiSeq run 
We have used the RIVM DADA2 pipeline to process all the sequence runs into an understandable count table format. Here we load each of the phyloseq objects (n = 11) generated by the pipeline into the R environment. 
```{r}
ps54 <- readRDS(file = "../Input/L054_VGO3_phyloseq.rds")
sample_names(ps54) <- gsub("_", "-", sample_names(ps54))
ps56 <- readRDS(file = "../Input/L056_VGO3_phyloseq.rds")
ps57 <- readRDS(file = "../Input/L057_VGO3_phyloseq.rds")
ps58 <- readRDS(file = "../Input/L058_VGO3_phyloseq.rds")
ps59 <- readRDS(file = "../Input/L059_VGO3_phyloseq.rds")
ps60 <- readRDS(file = "../Input/L060_VGO3_phyloseq.rds")
ps61 <- readRDS(file = "../Input/L061_VGO3_phyloseq.rds")
ps62 <- readRDS(file = "../Input/L062_VGO3_phyloseq.rds")
ps63 <- readRDS(file = "../Input/L063_VGO3_phyloseq.rds")
ps64 <- readRDS(file = "../Input/L064_VGO3_phyloseq.rds")
ps65 <- readRDS(file = "../Input/L065_VGO3_phyloseq.rds")

# read in the metadata, transform columns to numeric data type
meta <- read.csv("../Output/Metadata/VGO3_metadata_all_livestock_16Slab.csv", dec = ".", sep = ",") %>% 
  mutate(across(c(ct.value, X16S.QPCR.pg.ul, Picogreen.ug.ul), as.numeric))
# change the assay sample column (replace underscores with hyphens)
meta$assay_sample <- gsub("_", "-", meta$assay_sample)

# Create the phyloseq
gc() # clear unused memory

ps1 <- merge_phyloseq(ps54,ps56,ps57,ps58,ps59,ps60,ps61,ps62,ps63,ps64,ps65)
meta[!meta$assay_sample %in% sample_names(ps1),] # Check if there are any samples in the metadata that are not present in the ps1 object

ps1

# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 26480 taxa and 5572 samples ]
# sample_data() Sample Data:       [ 5572 samples by 11 sample variables ]
# tax_table()   Taxonomy Table:    [ 26480 taxa by 7 taxonomic ranks ]
```

We remove the "samples" that have less than 100 reads. These "samples" are likely artefacts generated by a mismatch in either one of the barcode pairs. 

```{r}
ps1@sam_data$read_nr <- rowSums(ps1@otu_table)
ps1 <- prune_samples(ps1@sam_data$read_nr > 100, ps1)
ps1 <- prune_samples(sample_names(ps1) != "R1-001", ps1)
ps1 <- prune_samples(sample_names(ps1) != "R1_001", ps1)
ps1

# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 26480 taxa and 2809 samples ]
# sample_data() Sample Data:       [ 2809 samples by 12 sample variables ]
# tax_table()   Taxonomy Table:    [ 26480 taxa by 7 taxonomic ranks ]
```
## Add metadata to the phyloseq object
```{r}
meta <- meta[meta$sample_id != "0",] # exclude rows where the sample_id is equal to "0".
sample_names(ps1) %in% meta$assay_sample %>% table() # check whether each sample name in ps1 exists in the assay_sample column of the meta dataframe.

m <- meta_to_df(ps1) # convert ps1 metadata to data frame
m$assay_sample <- m$sample_id # create a new column assay_sample in m by copying the values from sample_id
meta[!meta$assay_sample %in% m$assay_sample,] # Identify missing Samples
m <- m %>% select(-c(X, sample_identifier, Filtered, denoised_R1, denoised_R2,
                     Merged, non.chim, non.chim...filtered.by.length, Sample)) # remove unecessary columns
colnames(m) <- c("sample_id", "input_reads_nr", "perc_readds_kept_after_DADA2", "after_DADA2_reads_nr", "assay_sample") # Rename columns

m <- left_join(m, meta, by = "assay_sample") # merge ps1 metadata with all metadata
m <- m %>% rename(sample_id = sample_id.x) %>% select(-sample_id.y) # clean up any duplicate columns
colnames(m) <- gsub("X16S.QPCR.pg.ul", "qPCR", colnames(m)) # Change name
m$DI.Run <- as.character(m$DI.Run)
m <- subset(m, niche != "?NP") # remove rows where niche = ?NP
m$niche <- factor(m$niche, levels = c("NP", "OP", "BD", "BP", "FB", 
                                      "ZMCB", "ZMCD", "UMD", "BB"))
m$DI.Run <- factor(m$DI.Run, levels = m$DI.Run %>% unique() %>% as.numeric() %>% sort() %>% as.character())
```

We use the Zoo package to convert the date variables to actual dates.

```{r}
library(zoo)
m$visit_date <- as.Date(m$visit_date, format = "%Y-%m-%d")
m$DI.Date <- as.Date(m$DI.Date, origin = "1900-01-01")
m$pneumonia_lastyr_date <- as.Date(m$pneumonia_lastyr_date, format = "%Y-%m-%d")
m$acute_resp_infec_lastyr_date <- as.Date(m$acute_resp_infec_lastyr_date, format = "%Y-%m-%d")
m$covid19_confirmed_date <- as.Date(m$covid19_confirmed_date, format = "%Y-%m-%d")
m$covid19_suspected_date <- as.Date(m$covid19_suspected_date, format = "%Y-%m-%d")
m$covid19_vaccine1_date <- as.Date(m$covid19_vaccine1_date, format = "%Y-%m-%d")
m$covid19_vaccine2_date <- as.Date(m$covid19_vaccine2_date, format = "%Y-%m-%d")
m$antibiotics_lastyr_date <- as.Date(m$antibiotics_lastyr_date, format = "%Y-%m-%d")
m$antacids_lastyr_date <- as.Date(m$antacids_lastyr_date, format = "%Y-%m-%d")
m$covid19_vaccine3_date <- as.Date(m$covid19_vaccine3_date, format = "%Y-%m-%d")
# m$fuquestdate <- as.Date(m$fuquestdate, format = "%Y-%m-%d")
m$pneumonia_sympt_onset_date <- as.Date(m$pneumonia_sympt_onset_date, format = "%Y-%m-%d")
m$lamb_oldest_date <- as.Date(m$lamb_oldest_date, format = "%Y-%m-%d")
m$lamb_youngest_date <- as.Date(m$lamb_youngest_date, format = "%Y-%m-%d")
#m$special_work_date <- as.Date(m$special_work_date, format = "%Y-%m-%d")
#m$questionnaire_date <- as.Date(m$questionnaire_date, format = "%Y-%m-%d")
```

The total number of samples per population is as follows:

```{r}
#table(m$population, m$Miseq.Run, m$niche)
table(m$niche, m$population)
```

We also add a column for sampling_season and sampling_month using the visit_date information.

```{r}
m <- m %>% mutate(date1 =  as.Date(format(visit_date, '2021-%m-%d')), 
                  sampling_season = case_when(between(date1, ymd('2021-03-20'),
                                                      ymd('2021-06-19'))~ 'Spring',
                                              between(date1, ymd('2021-06-20'), 
                                                      ymd('2021-09-22'))~ 'Summer',
                                              between(date1, ymd('2021-09-22'), 
                                                      ymd('2021-12-21')) ~ 'Fall',
                                              TRUE ~ 'Winter'), date1 = NULL,
                  sampling_month = month(visit_date))
```

# Bacterial density & Picogreen

## Calculate the alpha diversity indices.

```{r}
m2 <- subset(m, niche != "NP?")
m2 <- subset(m2, niche != "BB")
```
Plots for qPCR

```{r qPCR_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$qPCR) + 
  labs(x = "Sample Type", y = "qPCR (pg/Âµl)") + 
  scale_y_continuous(trans='log10',
                     breaks=trans_breaks('log10', function(x) 10^x),
                     labels=trans_format('log10', math_format(10^.x))
                     ) +
    geom_hline(aes(yintercept = 0.095),  colour="grey50", linetype="dotted") + 
  facet_grid(population~Miseq.Run, scale = "fixed")+
  scale_color_manual(values = c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f"))
```

Plots for Picogreen

```{r Picogreen_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$Picogreen.ug.ul) + labs(x = "Sample Type", y = "Picogreen") + scale_y_continuous(trans='log10',
                     breaks=trans_breaks('log10', function(x) 10^x),
                     labels=trans_format('log10', math_format(10^.x))) + 
  facet_grid(population~Miseq.Run, scale = "fixed")+
  scale_color_manual(values = c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f"))
```

```{r reads_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$after_DADA2_reads_nr) + 
  labs(x = "Sample Type", y = "Number of reads") +
    geom_hline(aes(yintercept = 10000),  colour="grey50", linetype="dotted") + 
  facet_grid(population~Miseq.Run, scale = "fixed")+
  scale_color_manual(values = c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f"))
```

```{r reads_kept_raw, ow = '90%', h = 4, w = 7}
plot_sample_data(df=m2, x=m2$niche, y=m2$perc_readds_kept_after_DADA2) + 
  labs(x = "Sample Type", y = "% of reads kept") +
    geom_hline(aes(yintercept = 50),  colour="grey50", linetype="dotted") + 
  facet_grid(population~Miseq.Run, scale = "fixed")+
  scale_color_manual(values = c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f"))
```

# Filter and order the taxa

```{r}
ps1 <- prune_taxa(taxa_sums(ps1) > 0, ps1)
stats <- tibble(step = character(),
                samples = numeric(),
                taxa = numeric(),
                reads = numeric())
stats <- ps_stats(ps1, stats, "Merged phyloseq object")


ps1 <- taxa_sum_reorder(ps1) %>% separate_species()
taxa_names(ps1) <- gsub("_7_", "_", taxa_names(ps1))
```

```{r}
ps1 <- prune_taxa(taxa_sums(ps1) > 0, ps1)
stats <- ps_stats(ps1, stats, "Exclusion of Mitochondria, Chloroplast, Archaea & Eukaryota")
```

Add sample data to phyloseq object

```{r}
rownames(m) <- m$sample_id
m <- m %>% select(-c(assay_sample, sample_id))
sample_data(ps1) <- m
```

Save the phyloseq object:

```{r}
saveRDS(ps1, file = "../Output/Phyloseq/ps_raw.Rds")
```

# Session info
The session info is as follows:

```{r}
sessionInfo()
```


