---
title: "DA analysis"
author: "Beatrice Cornu Hewitt" 
date: "`r Sys.time()`"
---
# Packages
```{r}
library(phyloseq); library(dplyr); library(ANCOMBC); library(mia); library(DT); library(Maaslin2); library(microbiomer); library(patchwork);library(rcartocolor); library(ggplot2); library(scales); library(cartography); library(mia); library(rngtools); library(tidyverse); library(mia); library(car); library(DT); library(ANCOMBC); 

# Set the theme for ggplot
theme_set(theme_bw())
theme_update(
  axis.text.x = element_text(angle = 45, hjust = 1),
  strip.text = element_text(colour = 'white'),
  strip.background = element_rect(fill = "#2F4858", color = "#2F4858")
)
```
# Functions
```{r}
# Function to conduct MaAsLin2
do_maaslin2 <- function(ps, var, other_var, ref = NULL) { 
  #pre-processing
  sample_data(ps) <- sample_data(ps)[!is.na(sample_data(ps)[,var])]
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  otu <- ps@otu_table %>% as.data.frame() %>% t() %>% as.data.frame()
  meta <- meta_to_df(ps)
  rownames(meta) <- meta$sample_id
  meta <- as.data.frame(meta)
  otu <- otu[rownames(otu) %in% rownames(meta),]
  #Maaslin2
  ms = Maaslin2(
    input_data = otu, 
    input_metadata = meta, 
    output = var, 
    fixed_effects = c(var, other_var), 
    normalization = "NONE",
    transform = "LOG",
    min_abundance = 0,
    min_prevalence = 0,
    plot_heatmap = FALSE,
    plot_scatter = FALSE,
    reference = ref)
  #Edit dataframe
  ms <- ms$results
  ms$var <- var
  
  ms <- ms[ms$metadata == var,]
  
  ms <- ms %>% mutate(conf.low = coef-1.96*stderr,
                      conf.high = coef+1.96*stderr,
                      label = ifelse(qval < 0.1, "*", NA)
                      )
  return(ms)
}


Maaslin2:::LOG  # The internal MaAsLin log transformation takes the lowest species count and divides by 2 

```
# Read in data
## CP only
```{r}
# Read in the phyloseq object which has the new metadata and the RF model predictions
ps <- readRDS("../Output/Phyloseq/5_ps_complete_KRD_RFmodels_endotoxin.rds")
ps@sam_data$VGO3_ID[ps@sam_data$population=="GP"]
# Remove any samples from pneumonia patients as not used in the analysis
ps_cp <- subset_samples(ps, population == "CP")
ps_cp <- prune_taxa(taxa_sums(ps_cp) > 0, ps_cp) # Also remove taxa that are not in the CP ps object

# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_cp))$ASV)) # 17495
length(unique(as.data.frame(tax_table(ps_cp))$Genus)) # 960

# Replace NAs with 0s for each livestock variable
ps_cp@sam_data$KRD_nGoatsWghtDist.3000m.sum_std[is.na(ps_cp@sam_data$KRD_nGoatsWghtDist.3000m.sum_std)] <- 0
ps_cp@sam_data$KRD_nCattleFarm.3000m[is.na(ps_cp@sam_data$KRD_nCattleFarm.3000m)] <- 0
ps_cp@sam_data$KRD_nPigsWghtDist.3000m.sum_std[is.na(ps_cp@sam_data$KRD_nPigsWghtDist.3000m.sum_std)] <- 0
ps_cp@sam_data$KRD_nPoultryWghtDist.3000m.sum_std[is.na(ps_cp@sam_data$KRD_nPoultryWghtDist.3000m.sum_std)] <- 0

# Create new categorized variables of livestock variables based on tertiles

ps_cp@sam_data$Ecoli_Category <- cut(
  ps_cp@sam_data$ecoli_RF_preds_std,
  breaks = quantile(ps_cp@sam_data$ecoli_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$Staph_Category <- cut(
  ps_cp@sam_data$staph_RF_preds_std,
  breaks = quantile(ps_cp@sam_data$staph_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$tetW_Category <- cut(
  ps_cp@sam_data$tetw_RF_preds_std,
  breaks = quantile(ps_cp@sam_data$tetw_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$mecA_Category <- cut(
  ps_cp@sam_data$meca_RF_preds_std,
  breaks = quantile(ps_cp@sam_data$meca_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)


ps_cp@sam_data$Goats_Category <- cut(
  ps_cp@sam_data$KRD_nGoatsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp@sam_data$KRD_nGoatsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$Cattle_Category <- cut(
  ps_cp@sam_data$KRD_nCowsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp@sam_data$KRD_nCowsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$Pigs_Category <- cut(
  ps_cp@sam_data$KRD_nPigsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp@sam_data$KRD_nPigsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp@sam_data$Poultry_Category <- cut(
  ps_cp@sam_data$KRD_nPoultryWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp@sam_data$KRD_nPoultryWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)



# Extract data for visualization
data_to_plot <- data.frame(ps_cp@sam_data) %>%
  select(
    Ecoli_Category, Staph_Category, tetW_Category, mecA_Category,
    Goats_Category, Cattle_Category, Pigs_Category, Poultry_Category,
    ecoli_RF_preds_std, staph_RF_preds_std, tetw_RF_preds_std, meca_RF_preds_std,
    KRD_nGoatsWghtDist.3000m.sum_std, KRD_nCowsWghtDist.3000m.sum_std, 
    KRD_nPigsWghtDist.3000m.sum_std, KRD_nPoultryWghtDist.3000m.sum_std
  ) %>%
  pivot_longer(
    cols = c(ecoli_RF_preds_std, staph_RF_preds_std, tetw_RF_preds_std, meca_RF_preds_std,
             KRD_nGoatsWghtDist.3000m.sum_std, KRD_nCowsWghtDist.3000m.sum_std,
             KRD_nPigsWghtDist.3000m.sum_std, KRD_nPoultryWghtDist.3000m.sum_std),
    names_to = "variable",
    values_to = "value"
  ) %>%
  mutate(category = case_when(
    variable == "ecoli_RF_preds_std" ~ Ecoli_Category,
    variable == "staph_RF_preds_std" ~ Staph_Category,
    variable == "tetw_RF_preds_std" ~ tetW_Category,
    variable == "meca_RF_preds_std" ~ mecA_Category,
    variable == "KRD_nGoatsWghtDist.3000m.sum_std" ~ Goats_Category,
    variable == "KRD_nCowsWghtDist.3000m.sum_std" ~ Cattle_Category,
    variable == "KRD_nPigsWghtDist.3000m.sum_std" ~ Pigs_Category,
    variable == "KRD_nPoultryWghtDist.3000m.sum_std" ~ Poultry_Category
  ))

# Custom function to generate plots
create_tertile_plot <- function(data, var_name) {
  variable_data <- data %>% filter(variable == var_name)
  quantiles <- quantile(variable_data$value, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE)
  
  ggplot(variable_data, aes(x = value, fill = category)) +
    geom_histogram(bins = 30, alpha = 0.7, position = "identity", color = "black") +
    geom_vline(xintercept = quantiles, linetype = "dashed", color = "red") +
    scale_fill_manual(values = c("Low" = "skyblue", "Medium" = "gold", "High" = "tomato")) +
    theme_minimal() +
    labs(
      title = paste("Distribution of", var_name),
      x = "Value",
      y = "Frequency",
      fill = "Category"
    ) +
    annotate("text", x = quantiles, y = 0, label = round(quantiles, 2), color = "red", angle = 90, hjust = 1.2, vjust = 1.5, size = 3)
}

# List of variables to plot
variables_to_plot <- unique(data_to_plot$variable)

# Generate and save plots
for (var in variables_to_plot) {
  p <- create_tertile_plot(data_to_plot, var)
  ggsave(
    filename = paste0("../Output/Tertile_Plots/", var, "_tertile_plot.png"),
    plot = p,
    width = 8,
    height = 6
  )
}

```
## CP and GF 
```{r}
# CPs were matched 1:2 with the GFs (see other R notebook - "Matching_controls_goatfarmers.Rmd")
ps_cp_gf_matched <- readRDS("../Output/Phyloseq/3_ps_complete_matchedCPGF.rds")

# Count number of ASVs and genera in ps object
length(unique(as.data.frame(tax_table(ps_cp_gf_matched))$ASV)) # 7345
length(unique(as.data.frame(tax_table(ps_cp_gf_matched))$Genus)) # 761

# Replace NAs with 0s for each livestock variable for CP
ps_cp_gf_matched@sam_data$KRD_nGoatsWghtDist.3000m.sum_std[is.na(ps_cp_gf_matched@sam_data$KRD_nGoatsWghtDist.3000m.sum_std) & 
                                                      ps_cp_gf_matched@sam_data$population == "CP"] <- 0

ps_cp_gf_matched@sam_data$KRD_nCattleFarm.3000m[is.na(ps_cp_gf_matched@sam_data$KRD_nCattleFarm.3000m) & 
                                                      ps_cp_gf_matched@sam_data$population == "CP"] <- 0

ps_cp_gf_matched@sam_data$KRD_nPigsWghtDist.3000m.sum_std[is.na(ps_cp_gf_matched@sam_data$KRD_nPigsWghtDist.3000m.sum_std) & 
                                                      ps_cp_gf_matched@sam_data$population == "CP"] <- 0

ps_cp_gf_matched@sam_data$KRD_nPoultryWghtDist.3000m.sum_std[is.na(ps_cp_gf_matched@sam_data$KRD_nPoultryWghtDist.3000m.sum_std) & 
                                                      ps_cp_gf_matched@sam_data$population == "CP"] <- 0


# Create new categorized variables of livestock variables based on tertiles

ps_cp_gf_matched@sam_data$Ecoli_Category <- cut(
  ps_cp_gf_matched@sam_data$ecoli_RF_preds_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$ecoli_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$Staph_Category <- cut(
  ps_cp_gf_matched@sam_data$staph_RF_preds_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$staph_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$tetW_Category <- cut(
  ps_cp_gf_matched@sam_data$tetw_RF_preds_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$tetw_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$mecA_Category <- cut(
  ps_cp_gf_matched@sam_data$meca_RF_preds_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$meca_RF_preds_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)


ps_cp_gf_matched@sam_data$Goats_Category <- cut(
  ps_cp_gf_matched@sam_data$KRD_nGoatsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$KRD_nGoatsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$Cattle_Category <- cut(
  ps_cp_gf_matched@sam_data$KRD_nCowsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$KRD_nCowsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$Pigs_Category <- cut(
  ps_cp_gf_matched@sam_data$KRD_nPigsWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$KRD_nPigsWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)

ps_cp_gf_matched@sam_data$Poultry_Category <- cut(
  ps_cp_gf_matched@sam_data$KRD_nPoultryWghtDist.3000m.sum_std,
  breaks = quantile(ps_cp_gf_matched@sam_data$KRD_nPoultryWghtDist.3000m.sum_std, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE),
  labels = c("Low", "Medium", "High"),
  include.lowest = TRUE
)


```

## Check multicollinearity
```{r}
# Extract the sample data from the phyloseq object
sample_data_df <- data.frame(sample_data(ps_cp))
# Fit the linear model
lm_model <- lm(KRD_nGoatsWghtDist.3000m.sum_std ~ age + gender + sampling_season + smoked_ever +
                KRD_nCowsWghtDist.3000m.sum_std + KRD_nPoultryWghtDist.3000m.sum_std + 
                KRD_nPigsWghtDist.3000m.sum_std, data = sample_data_df)

# Calculate VIF values
vif_values <- vif(lm_model)

# Print the VIF values
print(vif_values) # All VIF values <2 therefore multicollinearity is not an issue


# Correlation plot between livestock variables
variables <- sample_data_df[, c("KRD_nPoultryWghtDist.3000m.sum_std", 
                                      "KRD_nCowsWghtDist.3000m.sum_std", 
                                      "KRD_nGoatsWghtDist.3000m.sum_std", 
                                      "KRD_nPigsWghtDist.3000m.sum_std")]

# Calculate the correlation matrix and p-values
corr_result <- psych::corr.test(variables)
corr_matrix <- corr_result$r    # Correlation matrix
p_matrix <- corr_result$p       # p-value matrix

# Convert correlation matrix and p-value matrix to long format
corr_long <- melt(corr_matrix)
p_long <- melt(p_matrix)

# Merge correlation and p-value matrices into one data frame
corr_long$p_value <- p_long$value

# Create labels based on significance levels
corr_long$label <- ifelse(corr_long$p_value < 0.001, "***",
                          ifelse(corr_long$p_value < 0.01, "**",
                                 ifelse(corr_long$p_value < 0.05, "*", "NS")))

# Only keep the upper triangle (excluding self-comparisons)
corr_long <- corr_long %>% 
  filter(as.numeric(Var1) < as.numeric(Var2))

# Plot the half-matrix heatmap with correlation coefficients and significance stars as labels
ggplot(corr_long, aes(Var1, Var2, fill = value)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(low = "#6D9EC1", high = "#E46726", mid = "white", 
                       midpoint = 0, limit = c(-1, 1), space = "Lab", 
                       name="Pearson\nCorrelation") +
  geom_text(aes(label = paste0(round(value, 2), "\n", label)), color = "black", size = 4) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.text.y = element_text(hjust = 1),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  ggtitle("Correlation Matrix with Significance Stars")

# High correlation between cattle and pigs



# Check distribution of animal variables 
hist(sample_data_df$KRD_nGoatsWghtDist.3000m.sum_std)
hist(sample_data_df$KRD_nCattleFarm.3000m)
hist(sample_data_df$KRD_nPigsWghtDist.3000m.sum_std)
hist(sample_data_df$KRD_nPoultryWghtDist.3000m.sum_std)



```

# Data processing
## Filtering and conversion to relative abundance
### CP only
```{r}
# Apply the presence abundance filter (Subramanian et al., Nature, 2014: 0.1% relative abundance in at least 2 samples - we do this for the overall data at first - rare taxa can introduce noise so first apply this global filter. 
# Then convert the ASVs to relative abundances to account for different sequencing depths (normalization step)
ps_cp_RA <- ps_cp %>% 
  pres_abund_filter() %>% 
  to_RA() 
# A total of 2215 ASVs were found to be present at or above a level of confident detection (0.1% relative abundance) in at least 2 samples (n = 15280 ASVs excluded).

# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_cp_RA))$ASV)) # 2215
length(unique(as.data.frame(tax_table(ps_cp_RA))$Genus)) # 314
```
### CP and GF
```{r}
ps_cp_gf_RA <- ps_cp_gf_matched %>% 
  pres_abund_filter() %>% 
  to_RA() 
# A total of 1305 ASVs were found to be present at or above a level of confident detection (0.1% relative abundance) in at least 2 samples (n = 7549 ASVs excluded).

# Count number of genera left in ps object
length(unique(as.data.frame(tax_table(ps_cp_gf_RA))$ASV)) # 1305
length(unique(as.data.frame(tax_table(ps_cp_gf_RA))$Genus)) # 266
```

## Convert to absolute abundances 
### CP only
```{r}
# Convert RA of ASVs by multiplying by the 16S qPCR measurements
cp_otu_tb <- otu_tab_to_df(ps_cp_RA)
cp_meta_tb <- meta_to_df(ps_cp_RA)
# Multiple RA by qPCR 
cp_otu_tb <- cp_otu_tb %>%
  mutate(across(-sample_id, ~.x*cp_meta_tb$qPCR)) %>% 
  as.data.frame()
# Make sample IDs the row names of the otu table
rownames(cp_otu_tb) <- cp_otu_tb$sample_id
# Remove the first column, transpose and convert to df
cp_otu_tb <- cp_otu_tb[,-1] %>% 
  t() %>%
  as.data.frame() 
# Create ps object with absolute abundance otu table
ps_abs_cp <- phyloseq(otu_table(cp_otu_tb, 
                             taxa_are_rows = T), 
                   tax_table(ps_cp_RA), 
                   sample_data(ps_cp_RA))
```
### CP and GF
```{r}
cp_gf_otu_tb <- otu_tab_to_df(ps_cp_gf_RA)
cp_gf_meta_tb <- meta_to_df(ps_cp_gf_RA)
# Multiple RA by qPCR 
cp_gf_otu_tb <- cp_gf_otu_tb %>%
  mutate(across(-sample_id, ~.x*cp_gf_meta_tb$qPCR)) %>% 
  as.data.frame()
# Make sample IDs the row names of the otu table
rownames(cp_gf_otu_tb) <- cp_gf_otu_tb$sample_id
# Remove the first column, transpose and convert to df
cp_gf_otu_tb <- cp_gf_otu_tb[,-1] %>% 
  t() %>%
  as.data.frame() 
# Create ps object with absolute abundance otu table
ps_abs_cp_gf <- phyloseq(otu_table(cp_gf_otu_tb, 
                             taxa_are_rows = T), 
                   tax_table(ps_cp_gf_RA), 
                   sample_data(ps_cp_gf_RA))
```
## Agglomerate data to genus level
### CP only
```{r}
ps_RA_genus_cp <- ps_abs_cp %>%
  tax_glom(taxrank = "Genus") %>% 
  to_RA() # 312 genera in the CP

# Change names to the genus-level names
taxa_names(ps_RA_genus_cp) <- ps_RA_genus_cp@tax_table[,6] 

# Convert relative abundances to absolute abundances by multiplying the ASV table with the qPCR results
cp_genus_otu_tb <- otu_tab_to_df(ps_RA_genus_cp)
cp_genus_meta_tb <- meta_to_df(ps_RA_genus_cp)

# Multiple RA by qPCR 
cp_genus_otu_tb <- cp_genus_otu_tb %>%
  mutate(across(-sample_id, ~.x*cp_genus_meta_tb$qPCR)) %>% 
  as.data.frame()

# Make sample IDs the row names of the otu table
rownames(cp_genus_otu_tb) <- cp_genus_otu_tb$sample_id
# Remove the first column, transpose and convert to df
cp_genus_otu_tb <- cp_genus_otu_tb[,-1] %>% 
  t() %>%
  as.data.frame() 

# Create ps object with absolute abundance otu table
ps_genus_abs_cp <- phyloseq(otu_table(cp_genus_otu_tb, 
                             taxa_are_rows = T), 
                   tax_table(ps_RA_genus_cp), 
                   sample_data(ps_RA_genus_cp))
```
### CP and GF
```{r}
ps_RA_genus_cp_gf <- ps_abs_cp_gf %>%
  tax_glom(taxrank = "Genus") %>% 
  to_RA() # 247 genera

# Change names to the genus-level names
taxa_names(ps_RA_genus_cp_gf) <- ps_RA_genus_cp_gf@tax_table[,6] 

# Convert relative abundances to absolute abundances by multiplying the ASV table with the qPCR results
cp_gf_genus_otu_tb <- otu_tab_to_df(ps_RA_genus_cp_gf)
cp_gf_genus_meta_tb <- meta_to_df(ps_RA_genus_cp_gf)

# Multiple RA by qPCR 
cp_gf_genus_otu_tb <- cp_gf_genus_otu_tb %>%
  mutate(across(-sample_id, ~.x*cp_gf_genus_meta_tb$qPCR)) %>% 
  as.data.frame()
# Make sample IDs the row names of the otu table
rownames(cp_gf_genus_otu_tb) <- cp_gf_genus_otu_tb$sample_id

# Remove the first column, transpose and convert to df
cp_gf_genus_otu_tb <- cp_gf_genus_otu_tb[,-1] %>% 
  t() %>%
  as.data.frame() 

# Create ps object with absolute abundance otu table
ps_genus_abs_cp_gf <- phyloseq(otu_table(cp_gf_genus_otu_tb, 
                             taxa_are_rows = T), 
                   tax_table(ps_RA_genus_cp_gf), 
                   sample_data(ps_RA_genus_cp_gf))
```

## Filter per niche 
### CP only 
#### ASV level 
##### NP 
```{r}
samples_cp_np <- sample_data(ps_abs_cp)$niche == "NP" 
# Prune the phyloseq object to include only NP samples
ps_abs_cp_np <- prune_samples(samples_cp_np, ps_abs_cp)
# Set up a prevalence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_cp_np <- nrow(ps_abs_cp_np@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_abs_cp_np_filt <- ps_abs_cp_np %>%
  pres_abund_filter(pres = filter_cp_np, abund = 0.005)
# A total of 15 ASVs were found to be present at or above a level of confident detection (0.5% relative abundance) in at least 83.6 samples (n = 2197 ASVs excluded).

# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_abs_cp_np_filt))$ASV)) # 15 ASVs
```
##### OP 
```{r}
# OROPHARYNX (OP)
samples_cp_op <- sample_data(ps_abs_cp)$niche == "OP" 
# Prune the phyloseq object to include only CP and OP samples
ps_abs_cp_op <- prune_samples(samples_cp_op, ps_abs_cp)
# Set up a prevalence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_cp_op <- nrow(ps_abs_cp_op@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_abs_cp_op_filt <- ps_abs_cp_op %>%
  pres_abund_filter(pres = filter_cp_op, abund = 0.01)
# A total of 40 ASVs were found to be present at or above a level of confident detection (1% relative abundance) in at least 97.8 samples (n = 2172 ASVs excluded).

# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_abs_cp_op_filt))$ASV)) # 40 ASVs
```
#### Genus level 
##### NP
```{r}
samples_cp_np <- sample_data(ps_genus_abs_cp)$niche == "NP" 
# Prune the phyloseq object to include only CP and NP samples
ps_abs_genus_cp_np <- prune_samples(samples_cp_np, ps_genus_abs_cp)

# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_cp_np_genus <- nrow(ps_abs_genus_cp_np@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_abs_genus_cp_np_filt <- ps_abs_genus_cp_np %>%
  pres_abund_filter(pres = filter_cp_np_genus, abund = 0.005)
# A total of 15 GENERA were found to be present at or above a level of confident detection (1% relative abundance) in at least 83.6 samples (n = 279 ASVs excluded).

# Count number of genera left in ps object
length(unique(as.data.frame(tax_table(ps_abs_genus_cp_np_filt))$Genus)) # 15 genera
```
##### OP
```{r}
samples_cp_op <- sample_data(ps_genus_abs_cp)$niche == "OP" 
# Prune the phyloseq object to include only CP and OP samples
ps_abs_genus_cp_op <- prune_samples(samples_cp_op, ps_genus_abs_cp)
# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_cp_op <- nrow(ps_abs_genus_cp_op@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_abs_genus_cp_op_filt <- ps_abs_genus_cp_op %>%
  pres_abund_filter(pres = filter_cp_op, abund = 0.01)
# A total of 21 GENERA were found to be present at or above a level of confident detection (1% relative abundance) in at least 97.8 samples (n = 268 ASVs excluded).

# Check number of genera remaining
length(unique(as.data.frame(tax_table(ps_abs_genus_cp_op_filt))$Genus)) # 21 genera
```
### CP and GF
#### ASV level 
##### NP 
```{r}
samples_cp_gf_np <- sample_data(ps_abs_cp_gf)$niche == "NP" 
# Prune the phyloseq object to include only CP and NP samples
ps_np_cp_gf_abs <- prune_samples(samples_cp_gf_np, ps_abs_cp_gf)
# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_np_cp_gf <- nrow(ps_np_cp_gf_abs@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_np_cp_gf_abs_filt <- ps_np_cp_gf_abs %>%
  pres_abund_filter(pres = filter_np_cp_gf, abund = 0.005)
# A total of 11 ASVs were found to be present at or above a level of confident detection (1% relative abundance) in at least 23.7 samples (n = 1292 ASVs excluded).


# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_np_cp_gf_abs_filt))$ASV)) # 17 ASVs
```
##### OP 
```{r}
samples_cp_gf_op <- sample_data(ps_abs_cp_gf)$niche == "OP" 
# Prune the phyloseq object to include only CP and OP samples
ps_op_cp_gf_abs <- prune_samples(samples_cp_gf_op, ps_abs_cp_gf)
# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_op_cp_gf <- nrow(ps_op_cp_gf_abs@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_op_cp_gf_abs_filt <- ps_op_cp_gf_abs %>%
  pres_abund_filter(pres = filter_op_cp_gf, abund = 0.01)
# A total of 40 ASVs were found to be present at or above a level of confident detection (1% relative abundance) in at least 28.7 samples (n = 1263 ASVs excluded).

# Count number of ASVs and genera left in ps object
length(unique(as.data.frame(tax_table(ps_op_cp_gf_abs_filt))$ASV)) # 40 ASVs
```
#### Genus level 
##### NP
```{r}
samples_cp_gf_np <- sample_data(ps_genus_abs_cp_gf)$niche == "NP"
# Prune the phyloseq object to include only CP and NP samples
ps_np_cp_gf_abs_genus <- prune_samples(samples_cp_gf_np, ps_genus_abs_cp_gf)
# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_np_cp_gf <- nrow(ps_np_cp_gf_abs_genus@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_np_cp_gf_abs_genus_filt <- ps_np_cp_gf_abs_genus %>%
  pres_abund_filter(pres = filter_np_cp_gf, abund = 0.005)

# A total of 11 ASVs were found to be present at or above a level of confident detection (1% relative abundance) in at least 23.7 samples (n = 231 ASVs excluded).

# Check number of genera remaining
length(unique(as.data.frame(tax_table(ps_np_cp_gf_abs_genus_filt))$Genus)) # 16 genera

```
##### OP
```{r}
samples_cp_gf_op <- sample_data(ps_genus_abs_cp_gf)$niche == "OP" 
# Prune the phyloseq object to include only CP and OP samples
ps_op_cp_gf_abs_genus <- prune_samples(samples_cp_gf_op, ps_genus_abs_cp_gf)

# Set up a presence abundance filter
# Keep samples that have a presence in at least 10% of all samples with an abundance of 0.01
# Determine what is 10% of samples
filter_op_cp_gf <- nrow(ps_op_cp_gf_abs_genus@sam_data) * 0.1 # 10% of total samples
# Apply the filter
ps_op_cp_gf_abs_genus_filt <- ps_op_cp_gf_abs_genus %>%
  pres_abund_filter(pres = filter_op_cp_gf, abund = 0.01)
# A total of 21 genera were found to be present at or above a level of confident detection (1% relative abundance) in at least 28.7 samples (n = 221 ASVs excluded).


# Check number of genera remaining
length(unique(as.data.frame(tax_table(ps_op_cp_gf_abs_genus_filt))$Genus)) # 21 genera
```

# Check metadata variables
```{r}
# MaAsLin fits a generalized linear model (GLM) for each feature (like microbial abundance) and metadata variable while adjusting for confounders and covariates. Here are the key assumptions and considerations for MaAsLin:
# 1. Independent Observations (e.g., no temporal or spatial autocorrelation).
# 2. Linearity (for linear models) - MaAsLin primarily models linear relationships between predictors (metadata) and outcomes (e.g., microbial abundance). Non-linear relationships between the metadata and microbial features may not be well captured.
# 3. Multicollinearity - metadata variables are not highly collinear with one another. 

# However MaAsLin can handle non-linear associations
# Select some example ASVs to plot
rownames(ps_abs_cp_np_filt@otu_table)
asvs_to_plot <- c("Dolosigranulum_pigrum_1", "Staphylococcus_2", "Corynebacterium_3")  # Replace with actual ASV names
metadata_vars <- c("KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", 
                   "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std", 
                   "ecoli_RF_preds_std", "staph_RF_preds_std", "tetw_RF_preds_std", "meca_RF_preds_std")

# Extract metadata and ASV abundance data from phyloseq object
metadata <- sample_data(ps_abs_cp_np_filt) %>% data.frame()
asv_data <- otu_table(ps_abs_cp_np_filt) %>% as.data.frame()

metadata$KRD_nGoatsWghtDist.3000m.sum_std
# Convert ASV data to long format and combine with metadata
asv_data_long <- asv_data %>% t  %>% as.data.frame()
rownames(asv_data_long) <- asv_data_long$VGO3_ID

# Merge metadata with ASV data
merged_data <- cbind(asv_data_long, metadata, by = "VGO3_ID")

# Filter the merged data to only include the selected ASVs and metadata variables
plot_data <- merged_data %>%
  select(c("VGO3_ID", asvs_to_plot, metadata_vars))

# Scatterplots
ggplot(plot_data, aes(x = ecoli_RF_preds_std, y = Dolosigranulum_pigrum_1)) +
  geom_point() +
  labs(x = "E.coli RF Predictions", y = "Dolosigranulum_pigrum_1 Abundance", 
       title = "Scatter Plot of E. coli exposure vs Dolosigranulum_pigrum_1") +
  theme_minimal()

# Scatterplots
ggplot(plot_data, aes(x = KRD_nGoatsWghtDist.3000m.sum_std, y = Staphylococcus_2)) +
  geom_point() +
  labs(x = "number of goats in 3km", y = "Staphylococcus_2 Abundance", 
       title = "Scatter Plot of KRD_nGoatsWghtDist.3000m.sum_std vs Staphylococcus_2") +
  theme_minimal()



# Histogram of ecoli_RF_preds_std
ggplot(plot_data, aes(x = ecoli_RF_preds_std)) +
  geom_histogram(binwidth = 0.5, fill = "blue", color = "black", alpha = 0.7) +
  labs(x = "E.coli RF Predictions", y = "Frequency", 
       title = "Histogram of E.coli RF Predictions") +
  theme_minimal()

ggplot(plot_data, aes(x = KRD_nGoatsWghtDist.3000m.sum_std)) +
  geom_histogram(binwidth = 0.5, fill = "blue", color = "black", alpha = 0.7) +
  labs(x = "Goats", y = "Frequency", 
       title = "Histogram of KRD_nGoatsWghtDist.3000m.sum_std") +
  theme_minimal()

ggplot(plot_data, aes(x = KRD_nCowsWghtDist.3000m.sum_std)) +
  geom_histogram(binwidth = 0.5, fill = "blue", color = "black", alpha = 0.7) +
  labs(x = "Cows", y = "Frequency", 
       title = "Histogram of KRD_nCowsWghtDist.3000m.sum_std") +
  theme_minimal()

ggplot(plot_data, aes(x = KRD_nPigsWghtDist.3000m.sum_std)) +
  geom_histogram(binwidth = 0.5, fill = "blue", color = "black", alpha = 0.7) +
  labs(x = "Pigs", y = "Frequency", 
       title = "Histogram of KRD_nPigsWghtDist.3000m.sum_std") +
  theme_minimal()

```

# MaAsLin2
## CP only
### ASV level 
#### NP niche
```{r}
# MULTIVARIABLE for livestock species
# Run MaAsLin for the (distance-based) livestock exposure variables (with adjustment for the other livestock variables)
# Goat farm exposure as main variable
Goat_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_np_filt, 
                               var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
                               other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                               ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
                               arrange(desc(coef))

# Save results as csv
write.csv(Goat_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/", "ASV_CP_NP_goat_std_continuous.csv"))

# Cattle farm exposure as main variable
Cow_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_np_filt, 
                               var = "KRD_nCowsWghtDist.3000m.sum_std",  # Including all livestock variables together
                               other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                               ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
                               arrange(desc(coef))
# Save results as csv
write.csv(Cow_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/", "ASV_CP_NP_cattle_std_continuous.csv"))

# Poultry farm exposure as main variable
Poultry_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_np_filt, 
                               var = "KRD_nPoultryWghtDist.3000m.sum_std",  # Including all livestock variables together
                               other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                               ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
                               arrange(desc(coef))
# Save results as csv
write.csv(Poultry_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/", "ASV_CP_NP_poultry_std_continuous.csv"))

# Pig farm exposure as main variable=
Pig_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_np_filt, 
                               var = "KRD_nPigsWghtDist.3000m.sum_std",  # Including all livestock variables together
                               other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nGoatsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                               ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
                               arrange(desc(coef))
# Save results as csv
write.csv(Pig_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/", "ASV_CP_NP_pig_std_continuous.csv"))


# Loop for individual models for modelled exposures: ecoli, staph, tetw, and meca
modelled_variables <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std","staph_RF_preds_std","tetw_RF_preds_std","meca_RF_preds_std")
# Create an empty list to store the ordered results
maaslin_results <- list()
# Loop and save each result to a CSV file
for (var in modelled_variables) {
  result <- do_maaslin2(ps = ps_abs_cp_np_filt,
                        var = var,
                        other_var = c("age", "gender", "sampling_season", "smoked_ever"), # adjust for these variables
                        ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>% # Set reference categories
                        arrange(desc(coef))
  # Save to a CSV file
  write.csv(result, paste0("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/", "ASV_CP_NP_", var, "_continuous.csv"))
}

```

##### Check model assumptions
```{r}
# Get the top 10 OTUs by abundance
top_otus <- names(sort(rowSums(ps_cp@otu_table), decreasing = TRUE)[1:10])
otu_top <- ps_cp@otu_table[top_otus,]  # Subset the top 10 OTUs

# Convert OTU data to a dataframe and transpose it (samples as rows, OTUs as columns)
otu_top_df <- as.data.frame(t(otu_top))  

# Extract metadata as a dataframe
meta_df <- meta_to_df(ps_cp)

meta_df <- data.frame(sample_data(ps_cp))
rownames(meta_df)

# Align rownames between OTU data and metadata
otu_top_df <- otu_top_df[rownames(otu_top_df) %in% rownames(meta_df), ]
meta_df <- meta_df[rownames(meta_df) %in% rownames(otu_top_df), ]

# Ensure rownames are in the same order
otu_top_df <- otu_top_df[match(rownames(meta_df), rownames(otu_top_df)), ]

# Combine OTU data and metadata (merge by column)
combined_df <- cbind(otu_top_df, meta_df)


# Select variables for correlation
cor_vars <- c("age", "gender", "sampling_season", "smoked_ever", 
              "KRD_nPigsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", 
              "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nGoatsWghtDist.3000m.sum_std")

# Convert categorical variables to numeric (one-hot encoding, if needed)
combined_df$gender <- as.numeric(as.factor(combined_df$gender))
combined_df$sampling_season <- as.numeric(as.factor(combined_df$sampling_season))
combined_df$smoked_ever <- as.numeric(as.factor(combined_df$smoked_ever))

# Extract OTU columns and the variables of interest
otu_cols <- colnames(otu_top_df)
cor_data <- combined_df[, c(otu_cols, cor_vars)]
# Compute correlation matrix
cor_matrix <- cor(cor_data, use = "pairwise.complete.obs")

# Extract correlations between OTUs and the selected variables
cor_otu_vars <- cor_matrix[otu_cols, cor_vars]
cor_otu_vars

# Initialize an empty dataframe to store results
cor_results <- data.frame()

# Loop through OTUs and variables
for (otu in otu_cols) {
  for (var in cor_vars) {
    test <- cor.test(combined_df[[otu]], combined_df[[var]], method = "pearson", use = "pairwise.complete.obs")
    cor_results <- rbind(cor_results, data.frame(
      OTU = otu,
      Variable = var,
      Correlation = test$estimate,
      p.value = test$p.value
    ))
  }
}

# Adjust p-values for multiple comparisons (optional)
cor_results$p.adjusted <- p.adjust(cor_results$p.value, method = "BH")
cor_results

library(ggplot2)
library(reshape2)

# Melt the correlation matrix for heatmap plotting
cor_melt <- melt(cor_otu_vars)

# Plot heatmap
ggplot(cor_melt, aes(x = Var2, y = Var1, fill = value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) +
  labs(x = "Variables", y = "OTUs", fill = "Correlation") +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


do_maaslin2_with_residuals <- function(ps, var, other_var, ref = NULL) { 
  # Preprocessing
  sample_data(ps) <- sample_data(ps)[!is.na(sample_data(ps)[,var])]
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  otu <- ps@otu_table %>% as.data.frame() %>% t() %>% as.data.frame()
  meta <- meta_to_df(ps)
  rownames(meta) <- meta$sample_id
  meta <- as.data.frame(meta)
  otu <- otu[rownames(otu) %in% rownames(meta),]
  
  # Run MaAsLin2
  ms = Maaslin2(
    input_data = otu, 
    input_metadata = meta, 
    output = var, 
    fixed_effects = c(var, other_var), 
    normalization = "NONE",
    min_abundance = 0,
    min_prevalence = 0,
    plot_heatmap = FALSE,
    plot_scatter = FALSE,
    reference = ref)
  
  # Edit dataframe
  ms <- ms$results
  ms$var <- var
  
  # Residual extraction
  residuals_list <- list()
  for (feature in unique(ms$feature)) {
    # Fit linear model for each feature
    formula <- as.formula(paste(feature, "~", paste(c(var, other_var), collapse = " + ")))
    lm_model <- lm(formula, data = cbind(otu, meta))
    
    # Store residuals
    residuals_list[[feature]] <- residuals(lm_model)
  }
  
  # Combine residuals into a dataframe
  residuals_df <- do.call(cbind, residuals_list)
  residuals_df <- as.data.frame(residuals_df)
  colnames(residuals_df) <- unique(ms$feature)
  
  # Return results with residuals
  return(list(results = ms, residuals = residuals_df))
}


Goat_expo_maaslin_res <- do_maaslin2_with_residuals(
  ps = ps_abs_cp_np_filt, 
  var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
  other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
  ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")  # Set reference categories
)

# Access the `results` data frame and arrange it
Goat_expo_results <- Goat_expo_maaslin_res$results %>% arrange(desc(coef))
# Access the `residuals` data frame
Goat_expo_residuals <- Goat_expo_maaslin_res$residuals


# Loop through the first 10 OTUs
for (i in 1:10) {
  residuals_otu <- Goat_expo_residuals[, i]
  otu_name <- colnames(Goat_expo_residuals)[i]
  
  # Plot residuals
  plot(
    residuals_otu, 
    main = paste("Residuals for", otu_name), 
    ylab = "Residuals", 
    xlab = "Index", 
    pch = 19, 
    col = "blue"
  )
  
  # Q-Q plot
  qqnorm(residuals_otu, main = paste("Q-Q Plot for", otu_name))
  qqline(residuals_otu, col = "red")
  
  # Histogram
  hist(
    residuals_otu, 
    breaks = 20, 
    main = paste("Histogram of Residuals for", otu_name), 
    xlab = "Residuals", 
    col = "lightgray", 
    border = "black"
  )
}


Goat_expo_maaslin_res$results

shapiro.test(Goat_expo_maaslin_res$residuals[,"Finegoldia_magna_52"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Lawsonella_clevelandensis_84"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Peptoniphilus_53"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Corynebacterium_3"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Anaerococcus_octavius_83"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Staphylococcus_2"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Dolosigranulum_pigrum_1"])
shapiro.test(Goat_expo_maaslin_res$residuals[,"Corynebacterium_67"])






ecoli_expo_maaslin_res <- do_maaslin2_with_residuals(
  ps = ps_abs_cp_np_filt, 
  var = "ecoli_RF_preds_std",  # Including all livestock variables together
  other_var = c("age", "gender", "sampling_season", "smoked_ever"),  # Adjusting for these variables
  ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")  # Set reference categories
)

# Access the `results` data frame and arrange it
ecoli_expo_maaslin_res_results <- ecoli_expo_maaslin_res$results %>% arrange(desc(coef))
# Access the `residuals` data frame
ecoli_expo_maaslin_res_residuals <- ecoli_expo_maaslin_res$residuals


# Loop through the first 10 OTUs
for (i in 1:10) {
  residuals_otu <- Goat_expo_residuals[, i]
  otu_name <- colnames(Goat_expo_residuals)[i]
  
  # Plot residuals
  plot(
    residuals_otu, 
    main = paste("Residuals for", otu_name), 
    ylab = "Residuals", 
    xlab = "Index", 
    pch = 19, 
    col = "blue"
  )
  
  # Q-Q plot
  qqnorm(residuals_otu, main = paste("Q-Q Plot for", otu_name))
  qqline(residuals_otu, col = "red")
  
  # Histogram
  hist(
    residuals_otu, 
    breaks = 20, 
    main = paste("Histogram of Residuals for", otu_name), 
    xlab = "Residuals", 
    col = "lightgray", 
    border = "black"
  )
}


shapiro.test(ecoli_expo_maaslin_res$residuals[,"Finegoldia_magna_52"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Lawsonella_clevelandensis_84"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Peptoniphilus_53"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Corynebacterium_3"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Anaerococcus_octavius_83"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Staphylococcus_2"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Dolosigranulum_pigrum_1"])
shapiro.test(ecoli_expo_maaslin_res$residuals[,"Corynebacterium_67"])














# Function to extract residuals and fitted values from MaAsLin2
do_maaslin2_with_residuals <- function(ps, var, other_var, ref = NULL) { 
  # Preprocessing
  sample_data(ps) <- sample_data(ps)[!is.na(sample_data(ps)[,var])]
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  otu <- ps@otu_table %>% as.data.frame() %>% t() %>% as.data.frame()
  meta <- meta_to_df(ps)
  rownames(meta) <- meta$sample_id
  meta <- as.data.frame(meta)
  otu <- otu[rownames(otu) %in% rownames(meta),]
  
  # Run MaAsLin2
  ms = Maaslin2(
    input_data = otu, 
    input_metadata = meta, 
    output = var, 
    fixed_effects = c(var, other_var), 
    normalization = "NONE",
    min_abundance = 0,
    min_prevalence = 0,
    plot_heatmap = FALSE,
    plot_scatter = FALSE,
    reference = ref)
  
  # Edit dataframe
  ms <- ms$results
  ms$var <- var
  
  # Residual and fitted values extraction
  residuals_list <- list()
  fitted_values_list <- list()
  for (feature in unique(ms$feature)) {
    # Fit linear model for each feature
    formula <- as.formula(paste(feature, "~", paste(c(var, other_var), collapse = " + ")))
    lm_model <- lm(formula, data = cbind(otu, meta))
    
    # Store residuals and fitted values
    residuals_list[[feature]] <- residuals(lm_model)
    fitted_values_list[[feature]] <- fitted(lm_model)
  }
  
  # Combine residuals and fitted values into dataframes
  residuals_df <- do.call(cbind, residuals_list)
  residuals_df <- as.data.frame(residuals_df)
  colnames(residuals_df) <- unique(ms$feature)
  
  fitted_values_df <- do.call(cbind, fitted_values_list)
  fitted_values_df <- as.data.frame(fitted_values_df)
  colnames(fitted_values_df) <- unique(ms$feature)
  
  # Return results with residuals and fitted values
  return(list(results = ms, residuals = residuals_df, fitted_values = fitted_values_df))
}

# Run the MaAsLin2 analysis and extract residuals and fitted values
Goat_expo_maaslin_res <- do_maaslin2_with_residuals(
  ps = ps_abs_cp_np_filt, 
  var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
  other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
  ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")  # Set reference categories
)

# Plotting the residuals vs fitted values for the first OTU (adjust for others as needed)
library(ggplot2)

# Create a plot for the first OTU (adjust to any specific OTU if needed)
otu_name1 <- colnames(Goat_expo_maaslin_res$residuals)[1]  # For the first OTU
# Get residuals and fitted values for this OTU
residuals_otu <- Goat_expo_maaslin_res$residuals[, otu_name1]
fitted_values_otu <- Goat_expo_maaslin_res$fitted_values[, otu_name1]
# Create a dataframe for plotting
residuals_fitted_df <- data.frame(
  Residuals = residuals_otu,
  Fitted_Values = fitted_values_otu
)
# Plot residuals vs fitted values
ggplot(residuals_fitted_df, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  labs(title = paste("Residuals vs Fitted Values for", otu_name),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()+
  ylim(-1, 1)  # Set the maximum y-value to 1, allow minimum to adjust automatically


# Create a plot for the first OTU (adjust to any specific OTU if needed)
otu_name2 <- colnames(Goat_expo_maaslin_res$residuals)[2]  # For the first OTU
# Get residuals and fitted values for this OTU
residuals_otu2 <- Goat_expo_maaslin_res$residuals[, otu_name2]
fitted_values_otu2 <- Goat_expo_maaslin_res$fitted_values[, otu_name2]
# Create a dataframe for plotting
residuals_fitted_df2 <- data.frame(
  Residuals = residuals_otu2,
  Fitted_Values = fitted_values_otu2
)
# Plot residuals vs fitted values
ggplot(residuals_fitted_df2, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  labs(title = paste("Residuals vs Fitted Values for", otu_name2),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()+
  ylim(-1, 1)  # Set the maximum y-value to 1, allow minimum to adjust automatically



# Create a plot for the first OTU (adjust to any specific OTU if needed)
otu_name3 <- colnames(Goat_expo_maaslin_res$residuals)[3]  # For the first OTU
# Get residuals and fitted values for this OTU
residuals_otu3 <- Goat_expo_maaslin_res$residuals[, otu_name3]
fitted_values_otu3 <- Goat_expo_maaslin_res$fitted_values[, otu_name3]
# Create a dataframe for plotting
residuals_fitted_df3 <- data.frame(
  Residuals = residuals_otu3,
  Fitted_Values = fitted_values_otu3
)
# Plot residuals vs fitted values
ggplot(residuals_fitted_df3, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  labs(title = paste("Residuals vs Fitted Values for", otu_name3),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()+
  ylim(-1, 1)  # Set the maximum y-value to 1, allow minimum to adjust automatically

# Fit a smoother line using geom_smooth() to confirm the pattern:
ggplot(residuals_fitted_df3, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  geom_smooth(method = "loess", color = "black") +  # Add a LOESS curve to visualize trends
  labs(title = paste("Residuals vs Fitted Values for", otu_name3),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()+
  ylim(-1, 1)  # Set the maximum y-value to 1, allow minimum to adjust automatically



library(dplyr)
residuals_fitted_df <- residuals_fitted_df %>%
  mutate(Fitted_Group = cut(Fitted_Values, breaks = quantile(Fitted_Values, probs = c(0, 0.25, 0.5, 0.75, 1)), include.lowest = TRUE))

residual_variance_by_group <- residuals_fitted_df %>%
  group_by(Fitted_Group) %>%
  summarise(Variance = var(Residuals))

print(residual_variance_by_group)




# Try log transform of predictor variables
ps_abs_cp_np_filt@sam_data$log_cows <- log1p(ps_abs_cp_np_filt@sam_data$KRD_nCowsWghtDist.3000m.sum_std)
ps_abs_cp_np_filt@sam_data$log_goats <- log1p(ps_abs_cp_np_filt@sam_data$KRD_nGoatsWghtDist.3000m.sum_std)
ps_abs_cp_np_filt@sam_data$log_poultry <- log1p(ps_abs_cp_np_filt@sam_data$KRD_nPoultryWghtDist.3000m.sum_std )
ps_abs_cp_np_filt@sam_data$log_pigs <- log1p(ps_abs_cp_np_filt@sam_data$KRD_nPigsWghtDist.3000m.sum_std)

# Repeat with log transformation
# Run the MaAsLin2 analysis and extract residuals and fitted values
Goat_expo_maaslin_res_log <- do_maaslin2_with_residuals(
  ps = ps_abs_cp_np_filt, 
  var = "log_goats",  # Including all livestock variables together
  other_var = c("age", "gender", "sampling_season", "smoked_ever", "log_cows", "log_poultry", "log_pigs"),  # Adjusting for these variables
  ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")  # Set reference categories
)

# Plotting the residuals vs fitted values for the first OTU (adjust for others as needed)
library(ggplot2)

# Create a plot for the first OTU (adjust to any specific OTU if needed)
otu_name <- colnames(Goat_expo_maaslin_res_log$residuals)[1]  # For the first OTU

# Get residuals and fitted values for this OTU
residuals_otu <- Goat_expo_maaslin_res_log$residuals[, otu_name]
fitted_values_otu <- Goat_expo_maaslin_res_log$fitted_values[, otu_name]

# Create a dataframe for plotting
residuals_fitted_df <- data.frame(
  Residuals = residuals_otu,
  Fitted_Values = fitted_values_otu
)

# Plot residuals vs fitted values
ggplot(residuals_fitted_df, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  labs(title = paste("Residuals vs Fitted Values for", otu_name),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()


residuals_fitted_df <- residuals_fitted_df %>%
  mutate(Fitted_Group = cut(Fitted_Values, breaks = quantile(Fitted_Values, probs = c(0, 0.25, 0.5, 0.75, 1)), include.lowest = TRUE))

residual_variance_by_group <- residuals_fitted_df %>%
  group_by(Fitted_Group) %>%
  summarise(Variance = var(Residuals))

print(residual_variance_by_group)







# Repeat with log transformation
ps_abs_cp_np_filt@sam_data$Pigs_Category
# Run the MaAsLin2 analysis and extract residuals and fitted values
Goat_expo_maaslin_res_categories <- do_maaslin2_with_residuals(
  ps = ps_abs_cp_np_filt, 
  var = "Goats_Category",  # Including all livestock variables together
  other_var = c("age", "gender", "sampling_season", "smoked_ever", "Cattle_Category", "Poultry_Category", "Pigs_Category"),  # Adjusting for these variables
  ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0","Cattle_Category,Low", "Poultry_Category,Low", "Pigs_Category,Low")  # Set reference categories
)

write.csv(Goat_expo_maaslin_res_categories, "../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/ASV_CP_NP_goat_exp_categorised.csv")


# Plotting the residuals vs fitted values for the first OTU (adjust for others as needed)
# Create a plot for the first OTU (adjust to any specific OTU if needed)
otu_name <- colnames(Goat_expo_maaslin_res_categories$residuals)[1]  # For the first OTU
# Get residuals and fitted values for this OTU
residuals_otu_cat <- Goat_expo_maaslin_res_categories$residuals[, otu_name]
fitted_values_otu_cat <- Goat_expo_maaslin_res_categories$fitted_values[, otu_name]
# Create a dataframe for plotting
residuals_fitted_df <- data.frame(
  Residuals = residuals_otu_cat,
  Fitted_Values = fitted_values_otu_cat
)

# Plot residuals vs fitted values
ggplot(residuals_fitted_df, aes(x = Fitted_Values, y = Residuals)) +
  geom_point(alpha = 0.5, color = "blue") +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  labs(title = paste("Residuals vs Fitted Values for", otu_name),
       x = "Fitted Values",
       y = "Residuals") +
  theme_minimal()


residuals_fitted_df <- residuals_fitted_df %>%
  mutate(Fitted_Group = cut(Fitted_Values, breaks = quantile(Fitted_Values, probs = c(0, 0.25, 0.5, 0.75, 1)), include.lowest = TRUE))

residual_variance_by_group <- residuals_fitted_df %>%
  group_by(Fitted_Group) %>%
  summarise(Variance = var(Residuals))

print(residual_variance_by_group)
```


#### OP niche
```{r}
# MULTIVARIABLE for livestock species
# Run MaAsLin for the (distance-based) livestock exposure variables (with adjustment for the other livestock variables)
# Goat farm exposure as main variable
Goat_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_op_filt, 
                                 var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                 other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                 ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Goat_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/", "ASV_level_maaslin_cp_op_goat_std_continuous.csv"))

# Cattle farm exposure as main variable
Cow_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_op_filt, 
                                var = "KRD_nCowsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Cow_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/", "ASV_level_maaslin_cp_op_cattle_std_continuous.csv"))

# Poultry farm exposure as main variable
Poultry_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_op_filt, 
                                    var = "KRD_nPoultryWghtDist.3000m.sum_std",  # Including all livestock variables together
                                    other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                    ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Poultry_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/", "ASV_level_maaslin_cp_op_poultry_std_continuous.csv"))

# Pig farm exposure as main variable=
Pig_expo_maaslin <- do_maaslin2(ps = ps_abs_cp_op_filt, 
                                var = "KRD_nPigsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nGoatsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Pig_expo_maaslin, paste0("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/", "ASV_level_maaslin_cp_op_pig_std_continuous.csv"))


# Loop for individual models for modelled exposures: ecoli, staph, tetw, and meca
modelled_variables <- c("EU_ANNUALAVG_PM10_2020_std","ecoli_RF_preds_std","staph_RF_preds_std","tetw_RF_preds_std","meca_RF_preds_std")
# Create an empty list to store the ordered results
maaslin_results <- list()

# Loop and save each result to a CSV file
for (var in modelled_variables) {
  result <- do_maaslin2(ps = ps_abs_cp_op_filt,
                        var = var,
                        other_var = c("age", "gender", "sampling_season", "smoked_ever"), # adjust for these variables
                        ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>% # Set reference categories
    arrange(desc(coef))
  # Save to a CSV file
  write.csv(result, paste0("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/", "ASV_cp_op_", var, "_continuous.csv"))
}
```
### Genus level 
#### NP niche
```{r}
# MULTIVARIABLE for livestock species
# Run MaAsLin for the (distance-based) livestock exposure variables (with adjustment for the other livestock variables)
# Goat farm exposure as main variable
Goat_expo_maaslin_genus <- do_maaslin2(ps = ps_abs_genus_cp_np_filt, 
                                 var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                 other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                 ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Goat_expo_maaslin_genus, paste0("../Output/DA_analysis/MaAsLin/CP_NP/genus_level/", "Genus_level_maaslin_cp_np_goat_std_continuous.csv"))

# Cattle farm exposure as main variable
Cow_expo_maaslin_genus <- do_maaslin2(ps = ps_abs_genus_cp_np_filt, 
                                var = "KRD_nCowsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Cow_expo_maaslin_genus, paste0("../Output/DA_analysis/MaAsLin/CP_NP/genus_level/", "Genus_level_maaslin_cp_np_cattle_std_continuous.csv"))

# Poultry farm exposure as main variable
Poultry_expo_maaslin_genus <- do_maaslin2(ps = ps_abs_genus_cp_np_filt, 
                                    var = "KRD_nPoultryWghtDist.3000m.sum_std",  # Including all livestock variables together
                                    other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                    ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Poultry_expo_maaslin_genus, paste0("../Output/DA_analysis/MaAsLin/CP_NP/genus_level/", "Genus_level_maaslin_cp_np_poultryfarmexposure.csv"))

# Pig farm exposure as main variable=
Pig_expo_maaslin_genus <- do_maaslin2(ps = ps_abs_genus_cp_np_filt, 
                                var = "KRD_nPigsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nGoatsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Pig_expo_maaslin_genus, paste0("../Output/DA_analysis/MaAsLin/CP_NP/genus_level/", "Genus_level_maaslin_cp_np_pig_std_continuous.csv"))


# Loop for individual models for modelled exposures: ecoli, staph, tetw, and meca
modelled_variables <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std","staph_RF_preds_std","tetw_RF_preds_std","meca_RF_preds_std")

# Loop and save each result to a CSV file
for (var in modelled_variables) {
  result <- do_maaslin2(ps = ps_abs_genus_cp_np_filt,
                        var = var,
                        other_var = c("age", "gender", "sampling_season", "smoked_ever"), # adjust for these variables
                        ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>% # Set reference categories
    arrange(desc(coef))
  # Save to a CSV file
  write.csv(result, paste0("../Output/DA_analysis/MaAsLin/CP_NP/Genus_level/", "Genus_cp_np_", var, "_continuous.csv"))
}

```

#### OP niche
```{r}
# MULTIVARIABLE for livestock species
# Run MaAsLin for the (distance-based) livestock exposure variables (with adjustment for the other livestock variables)
# Goat farm exposure as main variable
Goat_expo_maaslin_genus_cp_op <- do_maaslin2(ps = ps_abs_genus_cp_op_filt, 
                                       var = "KRD_nGoatsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                       other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                       ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Goat_expo_maaslin_genus_cp_op, paste0("../Output/DA_analysis/MaAsLin/CP_OP/genus_level/", "Genus_level_maaslin_cp_op_goat_std_continuous.csv"))

# Cattle farm exposure as main variable
Cow_expo_maaslin_genus_cp_op <- do_maaslin2(ps = ps_abs_genus_cp_op_filt, 
                                      var = "KRD_nCowsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                      other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                      ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Cow_expo_maaslin_genus_cp_op, paste0("../Output/DA_analysis/MaAsLin/CP_OP/genus_level/", "Genus_level_maaslin_cp_op_cattle_std_continuous.csv"))

# Poultry farm exposure as main variable
Poultry_expo_maaslin_genus_cp_op <- do_maaslin2(ps = ps_abs_genus_cp_op_filt, 
                                          var = "KRD_nPoultryWghtDist.3000m.sum_std",  # Including all livestock variables together
                                          other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPigsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                          ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Poultry_expo_maaslin_genus_cp_op, paste0("../Output/DA_analysis/MaAsLin/CP_OP/genus_level/", "Genus_level_maaslin_cp_op_poultry_std_continuous.csv"))

# Pig farm exposure as main variable=
Pig_expo_maaslin_genus_cp_op <- do_maaslin2(ps = ps_abs_genus_cp_op_filt, 
                                      var = "KRD_nPigsWghtDist.3000m.sum_std",  # Including all livestock variables together
                                      other_var = c("age", "gender", "sampling_season", "smoked_ever", "KRD_nCowsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", "KRD_nGoatsWghtDist.3000m.sum_std"),  # Adjusting for these variables
                                      ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>%  # Set reference categories
  arrange(desc(coef))
# Save results as csv
write.csv(Pig_expo_maaslin_genus_cp_op, paste0("../Output/DA_analysis/MaAsLin/CP_OP/genus_level/", "Genus_level_maaslin_cp_op_pig_std_continuous.csv"))


# Loop for individual models for modelled exposures: ecoli, staph, tetw, and meca
modelled_variables <- c("EU_ANNUALAVG_PM10_2020_std","ecoli_RF_preds_std","staph_RF_preds_std","tetw_RF_preds_std","meca_RF_preds_std")

# Loop and save each result to a CSV file
for (var in modelled_variables) {
  result <- do_maaslin2(ps = ps_abs_genus_cp_op_filt,
                        var = var,
                        other_var = c("age", "gender", "sampling_season", "smoked_ever"), # adjust for these variables
                        ref = c("gender,0", "sampling_season,Summer", "smoked_ever,0")) %>% # Set reference categories
    arrange(desc(coef))
  # Save to a CSV file
  write.csv(result, paste0("../Output/DA_analysis/MaAsLin/CP_OP/Genus_level/", "Genus_cp_op_", var, "_continuous.csv"))
}

```
## CP and GF
### ASV level 
#### NP niche
```{r}
# Run Maaslin2 analysis with CP as reference category for population
maaslin_cp_gf_np <- do_maaslin2(ps = ps_np_cp_gf_abs_filt, 
                             var = "population", 
                             other_var = c("age", "gender", "sampling_season", "smoked_ever"), 
                             ref = c("population,CP", "gender,0", "sampling_season,Summer", "smoked_ever,0"))

# Reorder ASVs based on effect size
maaslin_cp_gf_np_ord <- maaslin_cp_gf_np %>%
  arrange(desc(coef))

write.csv(maaslin_cp_gf_np_ord, "../Output/DA_analysis/MaAsLin/CP_GF_NP/ASV_level/ASV_level_maaslin_cp_gf_np.csv")
```
#### OP niche
```{r}
# Run Maaslin2 analysis with CP as reference category for population
maaslin_cp_gf_op <- do_maaslin2(ps = ps_op_cp_gf_abs_filt, 
                             var = "population", 
                             other_var = c( "age", "gender", "sampling_season", "smoked_ever"), 
                             ref = c("population,CP", "gender,0", "sampling_season,Summer", "smoked_ever,0"))

# Reorder ASVs based on effect size
maaslin_cp_gf_op_ord <- maaslin_cp_gf_op %>%
  arrange(desc(coef))

write.csv(maaslin_cp_gf_op_ord, "../Output/DA_analysis/MaAsLin/CP_GF_OP/ASV_level/ASV_level_maaslin_cp_gf_op.csv")
```

### Genus level 
#### NP niche
```{r}
# Run Maaslin2 analysis with CP as reference category for population
maaslin_cp_gf_genus_np <- do_maaslin2(ps = ps_np_cp_gf_abs_genus_filt, 
                             var = "population", 
                             other_var = c("age", "gender", "sampling_season", "smoked_ever"), 
                             ref = c("population,CP", "gender,0", "sampling_season,Summer", "smoked_ever,0"))

# Reorder ASVs based on effect size
maaslin_cp_gf_genus_np_ord <- maaslin_cp_gf_genus_np %>%
  arrange(desc(coef))

write.csv(maaslin_cp_gf_genus_np_ord, "../Output/DA_analysis/MaAsLin/CP_GF_NP/Genus_level/maaslin_cp_gf_genus_np_ord.csv")
```
#### OP niche
```{r}
# Run Maaslin2 analysis with CP as reference category for population
maaslin_cp_gf_genus_op <- do_maaslin2(ps = ps_op_cp_gf_abs_genus_filt, 
                                      var = "population", 
                                      other_var = c("age", "gender", "sampling_season", "smoked_ever"), 
                                      ref = c("population,CP", "gender,0", "sampling_season,Summer", "smoked_ever,0"))

# Reorder ASVs based on effect size
maaslin_cp_gf_genus_op_ord <- maaslin_cp_gf_genus_op %>%
  arrange(desc(coef))

write.csv(maaslin_cp_gf_genus_op_ord, "../Output/DA_analysis/MaAsLin/CP_GF_OP/Genus_level/maaslin_cp_gf_genus_op_ord.csv")
```


# ANCOM-BC2 analysis 
## CP only 
### ASV level
#### NP
```{r}
# Make tree summarised experiment data
np_cp_asv_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_abs_cp_np_filt)

# Set reference categories
np_cp_asv_tse@colData <- as.data.frame(np_cp_asv_tse@colData) %>%
  mutate(
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age),
    EU_ANNUALAVG_PM10_2020_std= as.numeric(EU_ANNUALAVG_PM10_2020_std),
    ecoli_RF_preds_std = as.numeric(ecoli_RF_preds_std),
    staph_RF_preds_std = as.numeric(staph_RF_preds_std),
    tetw_RF_preds_std = as.numeric(tetw_RF_preds_std),
    meca_RF_preds_std = as.numeric(meca_RF_preds_std)
  ) %>%
  DataFrame()  


models_cp <- list(
  livestock = "KRD_nGoatsWghtDist.3000m.sum_std + KRD_nCowsWghtDist.3000m.sum_std + 
               KRD_nPoultryWghtDist.3000m.sum_std + KRD_nPigsWghtDist.3000m.sum_std + 
               age + gender + sampling_season + smoked_ever",
  endotoxin = "EU_ANNUALAVG_PM10_2020_std + age + gender + sampling_season + smoked_ever",
  ecoli = "ecoli_RF_preds_std + age + gender + sampling_season + smoked_ever",
  staph = "staph_RF_preds_std + age + gender + sampling_season + smoked_ever",
  tetw = "tetw_RF_preds_std + age + gender + sampling_season + smoked_ever",
  meca = "meca_RF_preds_std + age + gender + sampling_season + smoked_ever"
)

# Run ANCOM-BC2 analyses for all models
ancom_cp_np <- map(models_cp, ~ ancombc2(
  data = np_cp_asv_tse,
  fix_formula = .x,
  rand_formula = NULL,
  p_adj_method = "BH",
  prv_cut = 0,
  lib_cut = 0,
  struc_zero = FALSE,
  alpha = 0.05,
  global = FALSE,
  pairwise = FALSE
))

# Process and save results 
ancom_cp_np <- map(ancom_cp_np, ~ mutate_if(.x$res, is.numeric, round, 2))

# Define output paths
output_dir <- "../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level"
file_names <- c("livestock", "endotoxin", "ecoli", "staph", "tetw", "meca")

# Save results as CSV
walk2(ancom_cp_np, file_names, ~ write_csv(.x, file.path(output_dir, paste0("ancom_np_cp_asv_", .y, ".csv"))))

```
#### OP
```{r}
# Make tree summarised experiment data
op_cp_asv_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_abs_cp_op_filt)

# Set reference categories
op_cp_asv_tse@colData <- as.data.frame(op_cp_asv_tse@colData) %>%
  mutate(
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age),
    EU_ANNUALAVG_PM10_2020_std = as.numeric(EU_ANNUALAVG_PM10_2020_std),
    ecoli_RF_preds_std = as.numeric(ecoli_RF_preds_std),
    staph_RF_preds_std = as.numeric(staph_RF_preds_std),
    tetw_RF_preds_std = as.numeric(tetw_RF_preds_std),
    meca_RF_preds_std = as.numeric(meca_RF_preds_std)
  ) %>%
  DataFrame()  


# Run ANCOM-BC2 analyses for all models
ancom_cp_op <- map(models_cp, ~ ancombc2(
  data = op_cp_asv_tse,
  fix_formula = .x,
  rand_formula = NULL,
  p_adj_method = "BH",
  prv_cut = 0,
  lib_cut = 0,
  struc_zero = FALSE,
  alpha = 0.05,
  global = FALSE,
  pairwise = FALSE
))

# Process and save results 
ancom_cp_op <- map(ancom_cp_op, ~ mutate_if(.x$res, is.numeric, round, 2))

# Define output paths
output_dir <- "../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level"
file_names <- c("livestock","endotoxin", "ecoli", "staph", "tetw", "meca")

# Save results as CSV
walk2(ancom_cp_op, file_names, ~ write_csv(.x, file.path(output_dir, paste0("ancom_op_cp_asv_", .y, ".csv"))))

```
### Genus level
#### NP 
```{r}
# Make tree summarised experiment data
np_cp_genus_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_abs_genus_cp_np_filt)

np_cp_genus_tse@colData <- as.data.frame(np_cp_genus_tse@colData) %>%
  mutate(
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age),
    EU_ANNUALAVG_PM10_2020_std = as.numeric(EU_ANNUALAVG_PM10_2020_std),
    ecoli_RF_preds_std = as.numeric(ecoli_RF_preds_std),
    staph_RF_preds_std = as.numeric(staph_RF_preds_std),
    tetw_RF_preds_std = as.numeric(tetw_RF_preds_std),
    meca_RF_preds_std = as.numeric(meca_RF_preds_std)
  ) %>%
  DataFrame()  


# Run ANCOM-BC2 analyses for all models
ancom_cp_np_genus <- map(models_cp, ~ ancombc2(
  data = np_cp_genus_tse,
  fix_formula = .x,
  rand_formula = NULL,
  p_adj_method = "BH",
  prv_cut = 0,
  lib_cut = 0,
  struc_zero = FALSE,
  alpha = 0.05,
  global = FALSE,
  pairwise = FALSE
))

# Process and save results 
ancom_cp_np_genus <- map(ancom_cp_np_genus, ~ mutate_if(.x$res, is.numeric, round, 2))

# Define output paths
output_dir <- "../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level"
file_names <- c("livestock", "endotoxin","ecoli", "staph", "tetw", "meca")

# Save results as CSV
walk2(ancom_cp_np_genus, file_names, ~ write_csv(.x, file.path(output_dir, paste0("ancom_np_cp_genus_", .y, ".csv"))))

```
#### OP 
```{r}

# Make tree summarised experiment data
op_cp_genus_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_abs_genus_cp_op_filt)

# Set reference categories
op_cp_genus_tse@colData <- as.data.frame(op_cp_genus_tse@colData) %>%
  mutate(
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age),
    EU_ANNUALAVG_PM10_2020_std = as.numeric(EU_ANNUALAVG_PM10_2020_std),
    ecoli_RF_preds_std = as.numeric(ecoli_RF_preds_std),
    staph_RF_preds_std = as.numeric(staph_RF_preds_std),
    tetw_RF_preds_std = as.numeric(tetw_RF_preds_std),
    meca_RF_preds_std = as.numeric(meca_RF_preds_std)
  ) %>%
  DataFrame() 


# Run ANCOM-BC2 analyses for all models
ancom_cp_op <- map(models_cp, ~ ancombc2(
  data = op_cp_genus_tse,
  fix_formula = .x,
  rand_formula = NULL,
  p_adj_method = "BH",
  prv_cut = 0,
  lib_cut = 0,
  struc_zero = FALSE,
  alpha = 0.05,
  global = FALSE,
  pairwise = FALSE
))

# Process and save results 
ancom_cp_op <- map(ancom_cp_op, ~ mutate_if(.x$res, is.numeric, round, 2))

# Define output paths
output_dir <- "../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level"
file_names <- c("livestock","endotoxin" ,"ecoli", "staph", "tetw", "meca")

# Save results as CSV
walk2(ancom_cp_op, file_names, ~ write_csv(.x, file.path(output_dir, paste0("ancom_op_cp_genus_", .y, ".csv"))))

```


## GF and CP
### ASV level
#### NP 
```{r}
# ANCOM-BC2 incorporates a sensitivity analysis to assess the impact of different pseudocoutns on zero counts for each taxon - it involves adding a series of pseudo-counts (ranging from 0.01 to 0.5 in increments of 0.01) to the zero counts of each taxon. Linear regression models are then performed on the bias-corrected log abundance table using the different pseudo-counts. The sensitivity score for each taxon is calculated as the proportion of times that the p-value exceeds the specified significance level (alpha). If all p-values consistently show significance or nonsignificance across different pseudo-counts and are consistent with the results obtained without adding pseudo-counts to zero counts (using the complete data), then the taxon is considered not sensitive to the pseudo-count addition.

# Make tree summarised experiment data
np_cp_gf_asv_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_np_cp_gf_abs_filt)

# Specify reference categories  
# Set reference categories
np_cp_gf_asv_tse@colData <- as.data.frame(np_cp_gf_asv_tse@colData) %>%
  mutate(
    population = factor(population, levels = c("CP", "GF")),
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age)
  ) %>%
  DataFrame()  

# Run ANCOM-BC2 analysis
ancom_np_cp_gf_asv <- ancombc2(data = np_cp_gf_asv_tse, 
                         fix_formula = "population  + age + gender + sampling_season + smoked_ever",
                         rand_formula = NULL,
                         p_adj_method = "BH",
                         prv_cut = 0,
                         lib_cut = 0,
                         group = "population",
                         struc_zero = FALSE,
                         alpha = 0.05,
                         global = TRUE,
                         pairwise = TRUE)


# Primary analysis results - also gives results from other factors included in the formula e.g. age, gender etc. 
ancom_np_cp_gf_asv_res <- ancom_np_cp_gf_asv$res %>%
    mutate_if(is.numeric, function(x) round(x, 2))


# Interpreting the table - for each variable there are the following columns for each ASV e.g. for the population variable
## lfc_populationGF - log fold change - the log fold change of the taxon abundance between the GF and CP groups.
## se_populationGF - Standard error of the lfc estimate 
## W_populationGF - Wald statistic which is used to test the null hypothesis that the log fold change for a variable is equal to zero (i.e., no association).
## p_populationGF - he p-value for the log fold change between GF and CP groups.
## q_populationGF - the adjusted p-value after multiple testing correction (e.g., Benjamini-Hochberg method), controlling for the false discovery rate (FDR).
## diff_populationGF - Indicates whether the taxon is differentially abundant for a specific variable
## passed_ss_populationGF - Indicates whether the taxon passed the structural zero (ss) test for a given variable.

write.csv(ancom_np_cp_gf_asv_res, "../Output/DA_analysis/ANCOM-BC/CP_GF_NP/ASV_level/ancom_np_cp_gf_asv_res2.csv")

# # Filter ASVs with q-value < 0.1 for the "population" variable
# # We are willing to accept a higher false-positive rate to avoid missing biologically meaningful taxa
# # Filter ASVs with q-value < 0.1 for the "population" variable
# significant_ASVs_np <- ancom_np_primary %>%
#   filter(q_populationGF < 0.1) # 5 ASVs that are significantly higher or lower
# 
# significant_ASVs_np_passedss <- ancom_np_primary %>%
#   filter(q_populationGF < 0.1 & passed_ss_populationGF == TRUE) # 2 ASVs that was significantly different and passed the ss test

```

#### OP
```{r}
op_cp_gf_asv_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_op_cp_gf_abs_filt)

# Specify reference categories  
op_cp_gf_asv_tse@colData <- as.data.frame(op_cp_gf_asv_tse@colData) %>%
  mutate(
    population = factor(population, levels = c("CP", "GF")),
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age)
  ) %>%
  DataFrame() 

# Run ANCOM-BC2 analysis
ancom_op_cp_gf_asv <- ancombc2(data = op_cp_gf_asv_tse, 
                         fix_formula = "population  + age + gender + sampling_season + smoked_ever",
                         rand_formula = NULL,
                         p_adj_method = "BH",
                         prv_cut = 0,
                         lib_cut = 0,
                         group = "population",
                         struc_zero = FALSE,
                         alpha = 0.05,
                         global = TRUE,
                         pairwise = TRUE)

# Primary analysis results - also gives results from other factors included in the formula e.g. age, gender etc. 
ancom_op_cp_gf_asv_res <- ancom_op_cp_gf_asv$res %>%
    mutate_if(is.numeric, function(x) round(x, 2))

# Save results
write.csv(ancom_op_cp_gf_asv_res, "../Output/DA_analysis/ANCOM-BC/CP_GF_OP/ASV_level/ancom_op_cp_gf_asv_res.csv")
```

### Genus level 
#### NP 
```{r}
# Make tree summarised experiment data
np_cp_gf_genus_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_np_cp_gf_abs_genus_filt)

# Specify reference categories  
np_cp_gf_genus_tse@colData <- as.data.frame(np_cp_gf_genus_tse@colData) %>%
  mutate(
    population = factor(population, levels = c("CP", "GF")),
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    
    age = as.numeric(age)
  ) %>%
  DataFrame() 


# Run ANCOM-BC2 analysis
ancom_np_cp_gf_genus <- ancombc2(data = np_cp_gf_genus_tse, 
                         fix_formula = "population  + age + gender + sampling_season + smoked_ever",
                         rand_formula = NULL,
                         p_adj_method = "BH",
                         prv_cut = 0,
                         lib_cut = 0,
                         group = "population",
                         struc_zero = FALSE,
                         alpha = 0.05,
                         global = TRUE,
                         pairwise = TRUE)


# Primary analysis results - also gives results from other factors included in the formula e.g. age, gender etc. 
ancom_np_cp_gf_genus_res <- ancom_np_cp_gf_genus$res %>%
    mutate_if(is.numeric, function(x) round(x, 2))

# save the overall model results
write.csv(ancom_np_cp_gf_genus_res, "../Output/DA_analysis/ANCOM-BC/CP_GF_NP/Genus_level/ancom_np_cp_gf_genus.csv")
```
#### OP
```{r}
# Make tree summarised experiment data
op_cp_gf_genus_tse <- mia::makeTreeSummarizedExperimentFromPhyloseq(ps_op_cp_gf_abs_genus_filt)

# Specify reference categories  
op_cp_gf_genus_tse@colData <- as.data.frame(op_cp_gf_genus_tse@colData) %>%
  mutate(
    population = factor(population, levels = c("CP", "GF")),
    gender = factor(gender, levels = c("0", "1")),
    sampling_season = factor(sampling_season, levels = c("Summer", "Fall", "Winter", "Spring")),
    smoked_ever = factor(smoked_ever, levels = c("0", "1", "2")),
    age = as.numeric(age)
  ) %>%
  DataFrame() 

# Run ANCOM-BC2 analysis
ancom_op_cp_gf_genus <- ancombc2(data = op_cp_gf_genus_tse, 
                         fix_formula = "population  + age + gender + sampling_season + smoked_ever",
                         rand_formula = NULL,
                         p_adj_method = "BH",
                         prv_cut = 0,
                         lib_cut = 0,
                         group = "population",
                         struc_zero = FALSE,
                         alpha = 0.05,
                         global = TRUE,
                         pairwise = TRUE)

 # Primary analysis results - also gives results from other factors included in the formula e.g. age, gender etc. 
ancom_op_cp_gf_genus_res <- ancom_op_cp_gf_genus$res %>%
    mutate_if(is.numeric, function(x) round(x, 2))

# save the overall model results
write.csv(ancom_op_cp_gf_genus_res, "../Output/DA_analysis/ANCOM-BC/CP_GF_OP/Genus_level/ancom_op_cp_gf_genus_res.csv")
```

# Visualising
## Boxplots to check abundances
### ASV level - all ASVs
#### NP 
```{r}
# Convert phyloseq object to long format (for all ASVs)
ps_all_ASVs_cp_gf_np <- ps_np_cp_gf_abs_filt %>%
  phyloseq::psmelt() %>%
  group_by(Sample, ASV, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

# Find the lowest non-zero abundant ASV
lowest_nonzero_ASV <- ps_all_ASVs_cp_gf_np %>%
  filter(Abundance > 0) %>%
  arrange(Abundance) 

lowest_ASV <- lowest_nonzero_ASV$Abundance[1]
# The smallest non-zero value in your dataset divded by two 
smallest_non_zero_value_div2<- (lowest_ASV/2)

# Add half of the smallest non-zero value to all abundances to add a constant so we don't have issues with log of 0 
ps_all_ASVs_cp_gf_np <- ps_all_ASVs_cp_gf_np %>%
  mutate(Abundance_log = log10(Abundance + (smallest_non_zero_value_div2)))

# Visualize the data
ggplot(ps_all_ASVs_cp_gf_np, aes(x = Abundance_log)) +
  geom_histogram(bins = 30, fill = "blue", alpha = 0.7) +
  theme_minimal() +
  labs(title = "Distribution of Abundance (Log Scale)", x = "Log(Abundance)", y = "Count")


# Order ASVs alphabetically for consistent visualization
ps_all_ASVs_cp_gf_np$ASV <- factor(ps_all_ASVs_cp_gf_np$ASV, levels = sort(unique(ps_all_ASVs_cp_gf_np$ASV)))

# Create violin and boxplots for all ASVs across pnpulations
plot_all_ASVs_np <- ggplot(ps_all_ASVs_cp_gf_np, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) +  # Violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Boxplot inside violin
  facet_wrap(~ASV, scales = "free_y", ncol = 4) +  # Facet by ASV with free y-scale
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) +  # Custom fill colors
  labs(
    title = "Abundance of All ASVs Across Populations",
    x = "Population",
    y = "Absolute Abundance (log-transformed)"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "top",
    strip.text = element_text(size = 8)  # Reduce facet title size for better readability
  )

# Display the plot
print(plot_all_ASVs_np)

# Save the plot
ggsave("../Output/DA_analysis/All_ASVs_violinplots_np.svg",
       plot_all_ASVs_np, width = 16, height = 12)

```
#### OP 
```{r}
# Convert phyloseq object to long format (for all ASVs)
ps_all_ASVs_cp_gf_op <- ps_op_cp_gf_abs_filt %>%
  phyloseq::psmelt() %>%
  group_by(Sample, ASV, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

# Find the lowest non-zero abundant ASV
lowest_nonzero_ASV_op <- ps_all_ASVs_cp_gf_op %>%
  filter(Abundance > 0) %>%
  arrange(Abundance) 

lowest_ASV_op <- lowest_nonzero_ASV_op$Abundance[1]
# The smallest non-zero value in your dataset divded by two 
lowest_ASV_op_div2<- (lowest_ASV_op/2)

# Add half of the smallest non-zero value to all abundances to add a constant so we don't have issues with log of 0 
ps_all_ASVs_cp_gf_op <- ps_all_ASVs_cp_gf_op %>%
  mutate(Abundance_log = log10(Abundance + (lowest_ASV_op_div2)))

# Order ASVs alphabetically for consistent visualization
ps_all_ASVs_cp_gf_op$ASV <- factor(ps_all_ASVs_cp_gf_op$ASV, levels = sort(unique(ps_all_ASVs_cp_gf_op$ASV)))

# Create violin and boxplots for all ASVs across populations
plot_all_ASVs <- ggplot(ps_all_ASVs_cp_gf_op, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) +  # Violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Boxplot inside violin
  facet_wrap(~ASV, scales = "free_y", ncol = 4) +  # Facet by ASV with free y-scale
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) +  # Custom fill colors
  labs(
    title = "Abundance of All ASVs Across Populations",
    x = "Population",
    y = "Absolute Abundance (log-transformed)"
  ) +
  theme_minimal() +  
  theme(
    panel.background = element_rect(fill = "white", color = NA),  # White panel background
    plot.background = element_rect(fill = "white", color = NA),   # White plot background
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "top",
    strip.text = element_text(size = 8)  # Reduce facet title size for better readability
  )

# Display the plot
print(plot_all_ASVs)

# Save the plot
ggsave("../Output/DA_analysis/All_ASVs_violinplots_op.png",
       plot_all_ASVs, width = 20, height = 20)
```

### Significant ASVs from MaAsLin
#### ASV level 
##### NP
```{r}
# ASV level results 
maaslin_cp_gf_np_ord

# Filter for significant results (e.g., FDR < 0.05)
significant_ASVs_maaslin_cp_gf_np <- maaslin_cp_gf_np_ord %>%
  filter(qval < 0.05) %>%
  arrange(desc(coef))

# Print the significant genera for confirmation
print(significant_ASVs_maaslin_cp_gf_np$feature)

# Extract the ASV names
significant_ASV_names_cp_gf_np <- significant_ASVs_maaslin_cp_gf_np$feature

# Subset the original phyloseq object to only include significant ASVs
ps_ASV_significant_cp_gf_np <- prune_taxa(taxa_names(ps_np_cp_gf_abs_filt) %in% significant_ASV_names_cp_gf_np, ps_np_cp_gf_abs_filt)

# Create a data frame for visualization
# Convert phyloseq object to long format (requires `phyloseq` and `tidyverse`)
ps_df_maaslin_cp_gf_np <- ps_ASV_significant_cp_gf_np %>%
  phyloseq::psmelt() %>%
  filter(ASV %in% significant_ASV_names_cp_gf_np) %>%
  group_by(Sample, ASV, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

ps_df_maaslin_cp_gf_np <- ps_df_maaslin_cp_gf_np %>%
  mutate(Abundance_log = log10(Abundance + 1)) # Log-transform abundance to reduce skew

# Optional: Order the genera by effect size for better visualization
ps_df_maaslin_cp_gf_np$ASV <- factor(ps_df_maaslin_cp_gf_np$ASV, levels = significant_ASV_names_cp_gf_np)

# Create a violin and boxplot (inside) of abundance for significant genera across populations
ggplot(ps_df_maaslin_cp_gf_np, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) + # Add violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Add boxplot inside violin
  facet_wrap(~ASV, scales = "free_y") +
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) + # Customize colors
  labs(title = "Differentially Abundant Genera",
       x = "Population",
       y = "Absolute abundance (log transformed)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "top")

# Save the plot
ggsave("../Output/DA_analysis/MaAsLin/CP_GF_NP/ASV_level/Significant_ASVs_violinplots.svg",
       width = 12, height = 8)
```

#### Genus level 
##### CP vs GF (NP) genus
```{r}
# Read in the MaAsLin2 results
maaslin_results_cp_gf_np <- read.csv("../Output/DA_analysis/MaAsLin/CP_GF_NP/Genus_level/maaslin_cp_gf_genus_np_ord.csv")

# Filter for significant results (e.g., FDR < 0.05)
significant_genera_cp_gf_np <- maaslin_results_cp_gf_np %>%
  filter(qval < 0.05) %>%
  arrange(desc(coef))

# Print the significant genera for confirmation
print(significant_genera_cp_gf_np$feature)

# Extract the genus names
significant_genus_names_cp_gf_np <- significant_genera_cp_gf_np$feature

# Subset your original phyloseq object (or microbiome dataset) to only include significant genera
# Assuming you have a phyloseq object named `ps_np_cp_gf_abs_genus_filt`
ps_significant_cp_gf_np <- prune_taxa(taxa_names(ps_np_cp_gf_abs_genus_filt) %in% significant_genus_names_cp_gf_np, ps_np_cp_gf_abs_genus_filt)

# Create a data frame for visualization
# Convert phyloseq object to long format (requires `phyloseq` and `tidyverse`)
ps_df_cp_gf_np <- ps_significant_cp_gf_np %>%
  phyloseq::psmelt() %>%
  filter(Genus %in% significant_genus_names_cp_gf_np) %>%
  group_by(Sample, Genus, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

ps_df_cp_gf_np <- ps_df_cp_gf_np %>%
  mutate(Abundance_log = log10(Abundance + 1)) # Log-transform abundance to reduce skew

# Optional: Order the genera by effect size for better visualization
ps_df_cp_gf_np$Genus <- factor(ps_df_cp_gf_np$Genus, levels = significant_genus_names_cp_gf_np)

# Create a violin and boxplot (inside) of abundance for significant genera across populations
ggplot(ps_df_cp_gf_np, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) + # Add violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Add boxplot inside violin
  facet_wrap(~Genus, scales = "free_y") +
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) + # Customize colors
  labs(title = "Differentially Abundant Genera",
       x = "Population",
       y = "Absolute abundance (log transformed)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "top")

# Save the plot
ggsave("../Output/DA_analysis/MaAsLin/CP_GF_NP/Genus_level/Significant_genera_violinplots.svg",
       width = 12, height = 8)
```
### CP vs GF (OP) ASV level
```{r}
# Filter for significant results (e.g., FDR < 0.05)
significant_ASVs_cp_gf_op <- maaslin_cp_gf_op_ord %>%
  filter(qval < 0.05) %>%
  arrange(desc(coef))

# Print the significant genera for confirmation
print(significant_ASVs_cp_gf_op$feature)

# Extract the ASV names
significant_ASV_names_cp_gf_op <- significant_ASVs_cp_gf_op$feature

# Subset your original phyloseq object (or microbiome dataset) to only include significant genera
# Assuming you have a phyloseq object named `ps_np_cp_gf_abs_ASV_filt`
ps_significant_ASVs_cp_gf_op <- prune_taxa(taxa_names(ps_op_cp_gf_abs_filt) %in% significant_ASV_names_cp_gf_op, ps_op_cp_gf_abs_filt)

# Create a data frame for visualization
# Convert phyloseq object to long format (requires `phyloseq` and `tidyverse`)
ps_significant_ASVs_cp_gf_op <- ps_significant_ASVs_cp_gf_op %>%
  phyloseq::psmelt() %>%
  filter(ASV %in% significant_ASV_names_cp_gf_op) %>%
  group_by(Sample, ASV, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

ps_significant_ASVs_cp_gf_op <- ps_significant_ASVs_cp_gf_op %>%
  mutate(Abundance_log = log10(Abundance + 1)) # Log-transform abundance to reduce skew

# Optional: Order the genera by effect size for better visualization
ps_significant_ASVs_cp_gf_op$ASV <- factor(ps_significant_ASVs_cp_gf_op$ASV, levels = significant_ASV_names_cp_gf_op)

# Create a violin and boxplot (inside) of abundance for significant genera across populations
ggplot(ps_significant_ASVs_cp_gf_op, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) + # Add violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Add boxplot inside violin
  facet_wrap(~ASV, scales = "free_y") +
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) + # Customize colors
  labs(title = "Differentially Abundant Genera",
       x = "Population",
       y = "Absolute abundance (log transformed)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "top")

# Save the plot
ggsave("../Output/DA_analysis/MaAsLin/CP_GF_OP/ASV_level/Significant_ASVs_violinplots.svg",
       width = 12, height = 8)



```

### CP vs GF (OP) genus level
```{r}
# Filter for significant results (e.g., FDR < 0.05)
significant_genera_cp_gf_op <- maaslin_cp_gf_genus_op_ord %>%
  filter(qval < 0.05) %>%
  arrange(desc(coef))

# Print the significant genera for confirmation
print(significant_genera_cp_gf_op$feature)

# Extract the genus names
significant_genus_names_cp_gf_op <- significant_genera_cp_gf_op$feature

# Subset your original phyloseq object (or microbiome dataset) to only include significant genera
# Assuming you have a phyloseq object named `ps_np_cp_gf_abs_genus_filt`
ps_significant_cp_gf_op <- prune_taxa(taxa_names(ps_op_cp_gf_abs_genus_filt) %in% significant_genus_names_cp_gf_op, ps_op_cp_gf_abs_genus_filt)

# Create a data frame for visualization
# Convert phyloseq object to long format (requires `phyloseq` and `tidyverse`)
ps_df_cp_gf_op <- ps_significant_cp_gf_op %>%
  phyloseq::psmelt() %>%
  filter(Genus %in% significant_genus_names_cp_gf_op) %>%
  group_by(Sample, Genus, population) %>%
  summarize(Abundance = sum(Abundance), .groups = "drop")

ps_df_cp_gf_op <- ps_df_cp_gf_op %>%
  mutate(Abundance_log = log10(Abundance + 1)) # Log-transform abundance to reduce skew

# Optional: Order the genera by effect size for better visualization
ps_df_cp_gf_op$Genus <- factor(ps_df_cp_gf_op$Genus, levels = significant_genus_names_cp_gf_op)

# Create a violin and boxplot (inside) of abundance for significant genera across populations
ggplot(ps_df_cp_gf_op, aes(x = population, y = Abundance_log, fill = population)) +
  geom_violin(alpha = 0.5, scale = "width", trim = TRUE, color = NA) + # Add violin plot
  geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.7, color = "black") + # Add boxplot inside violin
  facet_wrap(~Genus, scales = "free_y") +
  scale_fill_manual(values = c("CP" = "#e3be6a", "GF" = "#77b6a4")) + # Customize colors
  labs(title = "Differentially Abundant Genera",
       x = "Population",
       y = "Absolute abundance (log transformed)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "top")

# Save the plot
ggsave("../Output/DA_analysis/MaAsLin/CP_GF_OP/Genus_level/Significant_genera_violinplots.svg",
       width = 12, height = 8)
```

## MaAsLin heatmaps
## CP only
### ASV level 
#### NP
```{r}
# Define the directory
np_folder <- "../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/"

# Read NP results
np_files_cp_asvlevel <- list.files(np_folder, pattern = "continuous", full.names = TRUE)
np_results_cp_asvlevel <- lapply(np_files_cp_asvlevel, read.csv)
np_results_cp_asvlevel <- bind_rows(np_results_cp_asvlevel) 

# Extract relevant columns for the heatmap
np_heatmap_data <- np_results_cp_asvlevel %>%
  select(feature, metadata, coef, label) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std", "staph_RF_preds_std", "tetw_RF_preds_std", 
                           "meca_RF_preds_std", "KRD_nPigsWghtDist.3000m.sum_std", 
                           "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", 
                           "KRD_nCowsWghtDist.3000m.sum_std")

# Reorder metadata variables 
np_heatmap_data$metadata <- factor(np_heatmap_data$metadata, levels = custom_metadata_order)

# Rename metadata variables
np_heatmap_data <- np_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                                  "EU_ANNUALAVG_PM10_2020_std" = "Endotoxin exposure",
                                  "ecoli_RF_preds_std" = "E. coli exposure",
                                  "staph_RF_preds_std" = "Staph spp. exposure",
                                  "tetw_RF_preds_std" = "tetW exposure",
                                  "meca_RF_preds_std" = "mecA exposure",
                                  "KRD_nPigsWghtDist.3000m.sum_std" = "Pigs",
                                  "KRD_nGoatsWghtDist.3000m.sum_std" = "Goats",
                                  "KRD_nPoultryWghtDist.3000m.sum_std" = "Chickens",
                                  "KRD_nCowsWghtDist.3000m.sum_std" = "Cattle"
  ))

# Create the heatmap
heatmap_cp_np_asvlevel <- ggplot(np_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_NP/ASV_level/Heatmap_Maaslin_results_cp_np_asvlevel.svg", heatmap_cp_np_asvlevel, width = 6, height = 8)
```

#### OP 
```{r}
# Define the directory
op_folder <- "../Output/DA_analysis/MaAsLin/CP_OP/ASV_level"

# Read OP results
op_files_cp_asvlevel <- list.files(op_folder, pattern = "op_", full.names = TRUE)
op_results_cp_asvlevel <- lapply(op_files_cp_asvlevel, read.csv)
op_results_cp_asvlevel <- bind_rows(op_results_cp_asvlevel) 

# Extract relevant columns for the heatmap
op_heatmap_data <- op_results_cp_asvlevel %>%
  select(feature, metadata, coef, label) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std", "staph_RF_preds_std", "tetw_RF_preds_std", 
                           "meca_RF_preds_std", "KRD_nPigsWghtDist.3000m.sum_std", 
                           "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", 
                           "KRD_nCowsWghtDist.3000m.sum_std")

# Reorder metadata variables 
op_heatmap_data$metadata <- factor(op_heatmap_data$metadata, levels = custom_metadata_order)

# Rename metadata variables
op_heatmap_data <- op_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                                  "EU_ANNUALAVG_PM10_2020_std" = "Endotoxin exposure",
                                  "ecoli_RF_preds_std" = "E. coli exposure",
                                  "staph_RF_preds_std" = "Staph spp. exposure",
                                  "tetw_RF_preds_std" = "tetW exposure",
                                  "meca_RF_preds_std" = "mecA exposure",
                                  "KRD_nPigsWghtDist.3000m.sum_std" = "Pigs",
                                  "KRD_nGoatsWghtDist.3000m.sum_std" = "Goats",
                                  "KRD_nPoultryWghtDist.3000m.sum_std" = "Chickens",
                                  "KRD_nCowsWghtDist.3000m.sum_std" = "Cattle"
  ))

# Create the heatmap
heatmap_cp_op_asvlevel <- ggplot(op_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
    ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_OP/ASV_level/Heatmap_Maaslin_results_cp_op_asvlevel_multivar.svg", heatmap_cp_op_asvlevel, width = 6, height = 8)
```

### Genus level 
#### NP 
```{r}
# Define the directory
np_folder_genus <- "../Output/DA_analysis/MaAsLin/CP_NP/Genus_level/"

# Read NP results
np_files_cp_genuslevel <- list.files(np_folder_genus, pattern = "np_", full.names = TRUE)
np_results_cp_genuslevel <- lapply(np_files_cp_genuslevel, read.csv)
np_results_cp_genuslevel <- bind_rows(np_results_cp_genuslevel) 

# Extract relevant columns for the heatmap
np_heatmap_data <- np_results_cp_genuslevel %>%
  select(feature, metadata, coef, label) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std", "staph_RF_preds_std", "tetw_RF_preds_std", 
                           "meca_RF_preds_std", "KRD_nPigsWghtDist.3000m.sum_std", 
                           "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", 
                           "KRD_nCowsWghtDist.3000m.sum_std")

# Reorder metadata variables 
np_heatmap_data$metadata <- factor(np_heatmap_data$metadata, levels = custom_metadata_order)

# Rename metadata variables
np_heatmap_data <- np_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                                  "EU_ANNUALAVG_PM10_2020_std" = "Endotoxin exposure",
                                  "ecoli_RF_preds_std" = "E. coli exposure",
                                  "staph_RF_preds_std" = "Staph spp. exposure",
                                  "tetw_RF_preds_std" = "tetW exposure",
                                  "meca_RF_preds_std" = "mecA exposure",
                                  "KRD_nPigsWghtDist.3000m.sum_std" = "Pigs",
                                  "KRD_nGoatsWghtDist.3000m.sum_std" = "Goats",
                                  "KRD_nPoultryWghtDist.3000m.sum_std" = "Chickens",
                                  "KRD_nCowsWghtDist.3000m.sum_std" = "Cattle"
  ))


# Create the heatmap
heatmap_cp_np_genuslevel <- ggplot(np_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
)+
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_NP/Genus_level/Heatmap_Maaslin_results_cp_np_genuslevel.svg", heatmap_cp_np_genuslevel, width=6, height=8)
```

#### OP 
```{r}
# Define the directory
op_folder_genus <- "../Output/DA_analysis/MaAsLin/CP_OP/Genus_level/"

# Read OP results
op_files_cp_genuslevel <- list.files(op_folder_genus, pattern = "op_", full.names = TRUE)
op_results_cp_genuslevel <- lapply(op_files_cp_genuslevel, read.csv)
op_results_cp_genuslevel <- bind_rows(op_results_cp_genuslevel) 

# Extract relevant columns for the heatmap
op_heatmap_data <- op_results_cp_genuslevel %>%
  select(feature, metadata, coef, label) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("EU_ANNUALAVG_PM10_2020_std", "ecoli_RF_preds_std", "staph_RF_preds_std", "tetw_RF_preds_std", 
                           "meca_RF_preds_std", "KRD_nPigsWghtDist.3000m.sum_std", 
                           "KRD_nGoatsWghtDist.3000m.sum_std", "KRD_nPoultryWghtDist.3000m.sum_std", 
                           "KRD_nCowsWghtDist.3000m.sum_std")

# Reorder metadata variables 
op_heatmap_data$metadata <- factor(op_heatmap_data$metadata, levels = custom_metadata_order)

# Rename metadata variables
op_heatmap_data <- op_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                                  "EU_ANNUALAVG_PM10_2020_std" = "Endotoxin exposure",
                                  "ecoli_RF_preds_std" = "E. coli exposure",
                                  "staph_RF_preds_std" = "Staph spp. exposure",
                                  "tetw_RF_preds_std" = "tetW exposure",
                                  "meca_RF_preds_std" = "mecA exposure",
                                  "KRD_nPigsWghtDist.3000m.sum_std" = "Pigs",
                                  "KRD_nGoatsWghtDist.3000m.sum_std" = "Goats",
                                  "KRD_nPoultryWghtDist.3000m.sum_std" = "Chickens",
                                  "KRD_nCowsWghtDist.3000m.sum_std" = "Cattle"
  ))

# Create the heatmap
heatmap_cp_op_genuslevel <- ggplot(op_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_OP/Genus_level/Heatmap_Maaslin_results_cp_op_genuslevel_multivar.svg", heatmap_cp_op_genuslevel,width = 6, height =8)
```
## CP and GF
### ASV level 
#### NP 
```{r}
# Extract relevant columns for the heatmap
np_cp_gf_heatmap_data <- maaslin_cp_gf_np_ord %>%
  select(feature, metadata, coef, label)

# Rename metadata variables
np_cp_gf_heatmap_data <- np_cp_gf_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                           "population" = "Population"))


# Create the heatmap
heatmap_cp_gf_np_asvlevel <- ggplot(np_cp_gf_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_GF_NP/ASV_level/Heatmap_Maaslin_results_cp_gf_np_asvlevel.svg", heatmap_cp_gf_np_asvlevel, width = 6, height=6)
```
#### OP
```{r}
# Extract relevant columns for the heatmap
op_cp_gf_heatmap_data <- maaslin_cp_gf_op_ord %>%
  select(feature, metadata, coef, label)

# Rename metadata variables
op_cp_gf_heatmap_data <- op_cp_gf_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                           "population" = "Population"))


# Create the heatmap
heatmap_cp_gf_op_asvlevel <- ggplot(op_cp_gf_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_GF_OP/ASV_level/Heatmap_Maaslin_results_cp_gf_op_asvlevel.svg", heatmap_cp_gf_op_asvlevel, width = 6, height=6)
```

### Genus level 
#### NP 
```{r}
# Extract relevant columns for the heatmap
np_cp_gf_genus_heatmap_data <- maaslin_cp_gf_genus_np_ord %>%
  select(feature, metadata, coef, label)
# Rename metadata variables
np_cp_gf_genus_heatmap_data <- np_cp_gf_genus_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata, 
                                  "population" = "Population"))

# Create the heatmap
heatmap_cp_gf_np_genuslevel <- ggplot(np_cp_gf_genus_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_GF_NP/Genus_level/Heatmap_Maaslin_results_cp_gf_np_genuslevel.svg", heatmap_cp_gf_np_genuslevel, width = 4, height=6)
```
#### OP
```{r}
# Extract relevant columns for the heatmap
op_cp_gf_genus_heatmap_data <- maaslin_cp_gf_genus_op_ord %>%
  select(feature, metadata, coef, label)

# Rename metadata variables
op_cp_gf_genus_heatmap_data <- op_cp_gf_genus_heatmap_data %>%
  mutate(metadata = dplyr::recode(metadata,
                           "population" = "Population"))


# Create the heatmap
heatmap_cp_gf_op_genuslevel <- ggplot(op_cp_gf_genus_heatmap_data, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label), color = "black", size = 5) +  # Add labels inside the cells
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

ggsave("../Output/DA_analysis/MaAsLin/CP_GF_OP/Genus_level/Heatmap_Maaslin_results_cp_gf_op_genuslevel.svg", heatmap_cp_gf_op_genuslevel, width = 4.3, height=6)
```

# ANCOM-BC2 heatmaps
## CP only 
### ASV level 
#### NP 
```{r}
# Read in livestock exposure ancom results
ancom_np_cp_asv_livestock <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_livestock.csv")
colnames(ancom_np_cp_asv_livestock) <- trimws(colnames(ancom_np_cp_asv_livestock))
ancom_np_cp_asv_livestock <- as.data.frame(ancom_np_cp_asv_livestock)
# Goats
ancom_np_cp_asv_goat <- ancom_np_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nGoatsWghtDist.3000m.sum_std", "q_KRD_nGoatsWghtDist.3000m.sum_std", "passed_ss_KRD_nGoatsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Goats",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Cattle
ancom_np_cp_asv_cattle <- ancom_np_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nCowsWghtDist.3000m.sum_std", "q_KRD_nCowsWghtDist.3000m.sum_std", "passed_ss_KRD_nCowsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Cattle",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Pigs
ancom_np_cp_asv_pigs <- ancom_np_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPigsWghtDist.3000m.sum_std", "q_KRD_nPigsWghtDist.3000m.sum_std", "passed_ss_KRD_nPigsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Pigs",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Poultry
ancom_np_cp_asv_poultry <- ancom_np_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPoultryWghtDist.3000m.sum_std", "q_KRD_nPoultryWghtDist.3000m.sum_std", "passed_ss_KRD_nPoultryWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Chickens",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_asv_endotoxin <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_endotoxin.csv")
colnames(ancom_np_cp_asv_endotoxin) <- trimws(colnames(ancom_np_cp_asv_endotoxin))
ancom_np_cp_asv_endotoxin <- as.data.frame(ancom_np_cp_asv_endotoxin)

# Endotoxin
ancom_np_cp_asv_endotoxin <- ancom_np_cp_asv_endotoxin %>%
  select(any_of(c("taxon", "lfc_EU_ANNUALAVG_PM10_2020_std", "q_EU_ANNUALAVG_PM10_2020_std", "passed_ss_EU_ANNUALAVG_PM10_2020_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Endotoxin exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_np_cp_asv_ecoli <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_ecoli.csv")
colnames(ancom_np_cp_asv_ecoli) <- trimws(colnames(ancom_np_cp_asv_ecoli))
ancom_np_cp_asv_ecoli <- as.data.frame(ancom_np_cp_asv_ecoli)

# E. coli
ancom_np_cp_asv_ecoli <- ancom_np_cp_asv_ecoli %>%
  select(any_of(c("taxon", "lfc_ecoli_RF_preds_std", "q_ecoli_RF_preds_std", "passed_ss_ecoli_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "E. coli exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_asv_staph <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_staph.csv")
colnames(ancom_np_cp_asv_staph) <- trimws(colnames(ancom_np_cp_asv_staph))
ancom_np_cp_asv_staph <- as.data.frame(ancom_np_cp_asv_staph)

# Staph
ancom_np_cp_asv_staph <- ancom_np_cp_asv_staph %>%
  select(any_of(c("taxon", "lfc_staph_RF_preds_std", "q_staph_RF_preds_std", "passed_ss_staph_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Staph exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_asv_tetw <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_tetw.csv")
colnames(ancom_np_cp_asv_tetw) <- trimws(colnames(ancom_np_cp_asv_tetw))
ancom_np_cp_asv_tetw <- as.data.frame(ancom_np_cp_asv_tetw)

# tetw
ancom_np_cp_asv_tetw <- ancom_np_cp_asv_tetw %>%
  select(any_of(c("taxon", "lfc_tetw_RF_preds_std", "q_tetw_RF_preds_std", "passed_ss_tetw_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "tetW exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_np_cp_asv_meca <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/ancom_np_cp_asv_meca.csv")
colnames(ancom_np_cp_asv_meca) <- trimws(colnames(ancom_np_cp_asv_meca))
ancom_np_cp_asv_meca <- as.data.frame(ancom_np_cp_asv_meca)

# mecA
ancom_np_cp_asv_meca <- ancom_np_cp_asv_meca %>%
  select(any_of(c("taxon", "lfc_meca_RF_preds_std", "q_meca_RF_preds_std", "passed_ss_meca_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "mecA exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )



# Bind all exposure dataframes into one dataframe
combined_ancom_np_cp_asv <- bind_rows(
  ancom_np_cp_asv_goat,
  ancom_np_cp_asv_cattle,
  ancom_np_cp_asv_pigs,
  ancom_np_cp_asv_poultry,
  ancom_np_cp_asv_endotoxin,
  ancom_np_cp_asv_ecoli,
  ancom_np_cp_asv_staph,
  ancom_np_cp_asv_tetw,
  ancom_np_cp_asv_meca
)


combined_ancom_np_cp_asv

# Prepare data for the heatmap
combined_ancom_np_cp_asv <- combined_ancom_np_cp_asv %>%
  select(feature, metadata, coef, label, passed_ss) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("Endotoxin exposure","E. coli exposure", "Staph exposure", "tetW exposure", "mecA exposure", "Pigs", "Goats", "Chickens", "Cattle")

# Reorder metadata variables 
combined_ancom_np_cp_asv$metadata <- factor(combined_ancom_np_cp_asv$metadata, levels = custom_metadata_order)

# Modify label colors based on passed_ss
combined_ancom_np_cp_asv <- combined_ancom_np_cp_asv %>%
  mutate(label_color = ifelse(passed_ss == TRUE, "black", "white"))

# Create the heatmap
ancom_heatmap_cp_np_asv <- ggplot(combined_ancom_np_cp_asv, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label, color = label_color), size = 5) +  # Add labels with color based on passed_ss
  scale_color_identity() +  # Use actual color values for the labels
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

# Save the heatmap
ggsave("../Output/DA_analysis/ANCOM-BC/CP_NP/ASV_level/Heatmap_cp_np_asv.svg", 
       ancom_heatmap_cp_np_asv, width = 6, height = 6)
```

#### OP
```{r}
# Read in livestock exposure ancom results
ancom_op_cp_asv_livestock <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_livestock.csv")
colnames(ancom_op_cp_asv_livestock) <- trimws(colnames(ancom_op_cp_asv_livestock))
ancom_op_cp_asv_livestock <- as.data.frame(ancom_op_cp_asv_livestock)
# Goats
ancom_op_cp_asv_goat <- ancom_op_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nGoatsWghtDist.3000m.sum_std", "q_KRD_nGoatsWghtDist.3000m.sum_std", "passed_ss_KRD_nGoatsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Goats",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Cattle
ancom_op_cp_asv_cattle <- ancom_op_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nCowsWghtDist.3000m.sum_std", "q_KRD_nCowsWghtDist.3000m.sum_std", "passed_ss_KRD_nCowsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Cattle",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Pigs
ancom_op_cp_asv_pigs <- ancom_op_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPigsWghtDist.3000m.sum_std", "q_KRD_nPigsWghtDist.3000m.sum_std", "passed_ss_KRD_nPigsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Pigs",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Poultry
ancom_op_cp_asv_poultry <- ancom_op_cp_asv_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPoultryWghtDist.3000m.sum_std", "q_KRD_nPoultryWghtDist.3000m.sum_std", "passed_ss_KRD_nPoultryWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Chickens",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_asv_endotoxin <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_endotoxin.csv")
colnames(ancom_op_cp_asv_endotoxin) <- trimws(colnames(ancom_op_cp_asv_endotoxin))
ancom_op_cp_asv_endotoxin <- as.data.frame(ancom_op_cp_asv_endotoxin)

# Endotoxin
ancom_op_cp_asv_endotoxin <- ancom_op_cp_asv_endotoxin %>%
  select(any_of(c("taxon", "lfc_EU_ANNUALAVG_PM10_2020_std", "q_EU_ANNUALAVG_PM10_2020_std", "passed_ss_EU_ANNUALAVG_PM10_2020_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Endotoxin exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_asv_ecoli <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_ecoli.csv")
colnames(ancom_op_cp_asv_ecoli) <- trimws(colnames(ancom_op_cp_asv_ecoli))
ancom_op_cp_asv_ecoli <- as.data.frame(ancom_op_cp_asv_ecoli)

# E. coli
ancom_op_cp_asv_ecoli <- ancom_op_cp_asv_ecoli %>%
  select(any_of(c("taxon", "lfc_ecoli_RF_preds_std", "q_ecoli_RF_preds_std", "passed_ss_ecoli_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "E. coli exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_asv_staph <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_staph.csv")
colnames(ancom_op_cp_asv_staph) <- trimws(colnames(ancom_op_cp_asv_staph))
ancom_op_cp_asv_staph <- as.data.frame(ancom_op_cp_asv_staph)

# Staph
ancom_op_cp_asv_staph <- ancom_op_cp_asv_staph %>%
  select(any_of(c("taxon", "lfc_staph_RF_preds_std", "q_staph_RF_preds_std", "passed_ss_staph_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Staph exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_asv_tetw <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_tetw.csv")
colnames(ancom_op_cp_asv_tetw) <- trimws(colnames(ancom_op_cp_asv_tetw))
ancom_op_cp_asv_tetw <- as.data.frame(ancom_op_cp_asv_tetw)

# tetw
ancom_op_cp_asv_tetw <- ancom_op_cp_asv_tetw %>%
  select(any_of(c("taxon", "lfc_tetw_RF_preds_std", "q_tetw_RF_preds_std", "passed_ss_tetw_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "tetW exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_op_cp_asv_meca <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ancom_op_cp_asv_meca.csv")
colnames(ancom_op_cp_asv_meca) <- trimws(colnames(ancom_op_cp_asv_meca))
ancom_op_cp_asv_meca <- as.data.frame(ancom_op_cp_asv_meca)

# mecA
ancom_op_cp_asv_meca <- ancom_op_cp_asv_meca %>%
  select(any_of(c("taxon", "lfc_meca_RF_preds_std", "q_meca_RF_preds_std", "passed_ss_meca_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "mecA exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )



# Bind all exposure dataframes into one dataframe
combined_ancom_op_cp_asv <- bind_rows(
  ancom_op_cp_asv_goat,
  ancom_op_cp_asv_cattle,
  ancom_op_cp_asv_pigs,
  ancom_op_cp_asv_poultry,
  ancom_op_cp_asv_endotoxin,
  ancom_op_cp_asv_ecoli,
  ancom_op_cp_asv_staph,
  ancom_op_cp_asv_tetw,
  ancom_op_cp_asv_meca
)


combined_ancom_op_cp_asv

# Prepare data for the heatmap
combined_ancom_op_cp_asv <- combined_ancom_op_cp_asv %>%
  select(feature, metadata, coef, label, passed_ss) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("Endotoxin exposure", "E. coli exposure", "Staph exposure", "tetW exposure", "mecA exposure", "Pigs", "Goats", "Chickens", "Cattle")

# Reorder metadata variables 
combined_ancom_op_cp_asv$metadata <- factor(combined_ancom_op_cp_asv$metadata, levels = custom_metadata_order)

# Modify label colors based on passed_ss
combined_ancom_op_cp_asv <- combined_ancom_op_cp_asv %>%
  mutate(label_color = ifelse(passed_ss == TRUE, "black", "white"))

# Create the heatmap
ancom_heatmap_cp_op_asv <- ggplot(combined_ancom_op_cp_asv, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label, color = label_color), size = 5) +  # Add labels with color based on passed_ss
  scale_color_identity() +  # Use actual color values for the labels
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

# Save the heatmap
ggsave("../Output/DA_analysis/ANCOM-BC/CP_OP/ASV_level/ANCOM_heatmap_cp_op_asv.svg", 
       ancom_heatmap_cp_op_asv, width = 6, height = 7)
```

### Genus level 
#### NP
```{r}
# Read in livestock exposure ancom results
ancom_np_cp_genus_livestock <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_livestock.csv")
colnames(ancom_np_cp_genus_livestock) <- trimws(colnames(ancom_np_cp_genus_livestock))
ancom_np_cp_genus_livestock <- as.data.frame(ancom_np_cp_genus_livestock)
# Goats
ancom_np_cp_genus_goat <- ancom_np_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nGoatsWghtDist.3000m.sum_std", "q_KRD_nGoatsWghtDist.3000m.sum_std", "passed_ss_KRD_nGoatsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Goats",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Cattle
ancom_np_cp_genus_cattle <- ancom_np_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nCowsWghtDist.3000m.sum_std", "q_KRD_nCowsWghtDist.3000m.sum_std", "passed_ss_KRD_nCowsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Cattle",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Pigs
ancom_np_cp_genus_pigs <- ancom_np_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPigsWghtDist.3000m.sum_std", "q_KRD_nPigsWghtDist.3000m.sum_std", "passed_ss_KRD_nPigsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Pigs",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Poultry
ancom_np_cp_genus_poultry <- ancom_np_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPoultryWghtDist.3000m.sum_std", "q_KRD_nPoultryWghtDist.3000m.sum_std", "passed_ss_KRD_nPoultryWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Chickens",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_genus_endotoxin <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_endotoxin.csv")
colnames(ancom_np_cp_genus_endotoxin) <- trimws(colnames(ancom_np_cp_genus_endotoxin))
ancom_np_cp_genus_endotoxin <- as.data.frame(ancom_np_cp_genus_endotoxin)

# Endotoxin
ancom_np_cp_genus_endotoxin <- ancom_np_cp_genus_endotoxin %>%
  select(any_of(c("taxon", "lfc_EU_ANNUALAVG_PM10_2020_std", "q_EU_ANNUALAVG_PM10_2020_std", "passed_ss_EU_ANNUALAVG_PM10_2020_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Endotoxin exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_np_cp_genus_ecoli <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_ecoli.csv")
colnames(ancom_np_cp_genus_ecoli) <- trimws(colnames(ancom_np_cp_genus_ecoli))
ancom_np_cp_genus_ecoli <- as.data.frame(ancom_np_cp_genus_ecoli)

# E. coli
ancom_np_cp_genus_ecoli <- ancom_np_cp_genus_ecoli %>%
  select(any_of(c("taxon", "lfc_ecoli_RF_preds_std", "q_ecoli_RF_preds_std", "passed_ss_ecoli_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "E. coli exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_genus_staph <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_staph.csv")
colnames(ancom_np_cp_genus_staph) <- trimws(colnames(ancom_np_cp_genus_staph))
ancom_np_cp_genus_staph <- as.data.frame(ancom_np_cp_genus_staph)

# Staph
ancom_np_cp_genus_staph <- ancom_np_cp_genus_staph %>%
  select(any_of(c("taxon", "lfc_staph_RF_preds_std", "q_staph_RF_preds_std", "passed_ss_staph_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Staph exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_np_cp_genus_tetw <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_tetw.csv")
colnames(ancom_np_cp_genus_tetw) <- trimws(colnames(ancom_np_cp_genus_tetw))
ancom_np_cp_genus_tetw <- as.data.frame(ancom_np_cp_genus_tetw)

# tetw
ancom_np_cp_genus_tetw <- ancom_np_cp_genus_tetw %>%
  select(any_of(c("taxon", "lfc_tetw_RF_preds_std", "q_tetw_RF_preds_std", "passed_ss_tetw_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "tetW exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_np_cp_genus_meca <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ancom_np_cp_genus_meca.csv")
colnames(ancom_np_cp_genus_meca) <- trimws(colnames(ancom_np_cp_genus_meca))
ancom_np_cp_genus_meca <- as.data.frame(ancom_np_cp_genus_meca)

# mecA
ancom_np_cp_genus_meca <- ancom_np_cp_genus_meca %>%
  select(any_of(c("taxon", "lfc_meca_RF_preds_std", "q_meca_RF_preds_std", "passed_ss_meca_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "mecA exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )



# Bind all exposure dataframes into one dataframe
combined_ancom_np_cp_genus <- bind_rows(
  ancom_np_cp_genus_goat,
  ancom_np_cp_genus_cattle,
  ancom_np_cp_genus_pigs,
  ancom_np_cp_genus_poultry,
  ancom_np_cp_genus_endotoxin,
  ancom_np_cp_genus_ecoli,
  ancom_np_cp_genus_staph,
  ancom_np_cp_genus_tetw,
  ancom_np_cp_genus_meca
)


combined_ancom_np_cp_genus

# Prepare data for the heatmap
combined_ancom_np_cp_genus <- combined_ancom_np_cp_genus %>%
  select(feature, metadata, coef, label, passed_ss) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("Endotoxin exposure", "E. coli exposure", "Staph exposure", "tetW exposure", "mecA exposure", "Pigs", "Goats", "Chickens", "Cattle")

# Reorder metadata variables 
combined_ancom_np_cp_genus$metadata <- factor(combined_ancom_np_cp_genus$metadata, levels = custom_metadata_order)

# Modify label colors based on passed_ss
combined_ancom_np_cp_genus <- combined_ancom_np_cp_genus %>%
  mutate(label_color = ifelse(passed_ss == TRUE, "black", "white"))

# Create the heatmap
ancom_heatmap_cp_np_genus <- ggplot(combined_ancom_np_cp_genus, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label, color = label_color), size = 5) +  # Add labels with color based on passed_ss
  scale_color_identity() +  # Use actual color values for the labels
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

# Save the heatmap
ggsave("../Output/DA_analysis/ANCOM-BC/CP_NP/Genus_level/ANCOM_heatmap_cp_np_genus.svg", 
       ancom_heatmap_cp_np_genus, width = 6, height = 7)
```



#### OP
```{r}
# Read in livestock exposure ancom results
ancom_op_cp_genus_livestock <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_livestock.csv")
colnames(ancom_op_cp_genus_livestock) <- trimws(colnames(ancom_op_cp_genus_livestock))
ancom_op_cp_genus_livestock <- as.data.frame(ancom_op_cp_genus_livestock)
# Goats
ancom_op_cp_genus_goat <- ancom_op_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nGoatsWghtDist.3000m.sum_std", "q_KRD_nGoatsWghtDist.3000m.sum_std", "passed_ss_KRD_nGoatsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Goats",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Cattle
ancom_op_cp_genus_cattle <- ancom_op_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nCowsWghtDist.3000m.sum_std", "q_KRD_nCowsWghtDist.3000m.sum_std", "passed_ss_KRD_nCowsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Cattle",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Pigs
ancom_op_cp_genus_pigs <- ancom_op_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPigsWghtDist.3000m.sum_std", "q_KRD_nPigsWghtDist.3000m.sum_std", "passed_ss_KRD_nPigsWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Pigs",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Poultry
ancom_op_cp_genus_poultry <- ancom_op_cp_genus_livestock %>%
  select(any_of(c("taxon", "lfc_KRD_nPoultryWghtDist.3000m.sum_std", "q_KRD_nPoultryWghtDist.3000m.sum_std", "passed_ss_KRD_nPoultryWghtDist.3000m.sum_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Chickens",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_genus_endotoxin <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_endotoxin.csv")
colnames(ancom_op_cp_genus_endotoxin) <- trimws(colnames(ancom_op_cp_genus_endotoxin))
ancom_op_cp_genus_endotoxin <- as.data.frame(ancom_op_cp_genus_endotoxin)

# Endotoxin
ancom_op_cp_genus_endotoxin <- ancom_op_cp_genus_endotoxin %>%
  select(any_of(c("taxon", "lfc_EU_ANNUALAVG_PM10_2020_std", "q_EU_ANNUALAVG_PM10_2020_std", "passed_ss_EU_ANNUALAVG_PM10_2020_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Endotoxin exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )
# Read in livestock exposure (modelled) ancom results
ancom_op_cp_genus_ecoli <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_ecoli.csv")
colnames(ancom_op_cp_genus_ecoli) <- trimws(colnames(ancom_op_cp_genus_ecoli))
ancom_op_cp_genus_ecoli <- as.data.frame(ancom_op_cp_genus_ecoli)

# E. coli
ancom_op_cp_genus_ecoli <- ancom_op_cp_genus_ecoli %>%
  select(any_of(c("taxon", "lfc_ecoli_RF_preds_std", "q_ecoli_RF_preds_std", "passed_ss_ecoli_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "E. coli exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_genus_staph <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_staph.csv")
colnames(ancom_op_cp_genus_staph) <- trimws(colnames(ancom_op_cp_genus_staph))
ancom_op_cp_genus_staph <- as.data.frame(ancom_op_cp_genus_staph)

# Staph
ancom_op_cp_genus_staph <- ancom_op_cp_genus_staph %>%
  select(any_of(c("taxon", "lfc_staph_RF_preds_std", "q_staph_RF_preds_std", "passed_ss_staph_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "Staph exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )

# Read in livestock exposure (modelled) ancom results
ancom_op_cp_genus_tetw <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_tetw.csv")
colnames(ancom_op_cp_genus_tetw) <- trimws(colnames(ancom_op_cp_genus_tetw))
ancom_op_cp_genus_tetw <- as.data.frame(ancom_op_cp_genus_tetw)

# tetw
ancom_op_cp_genus_tetw <- ancom_op_cp_genus_tetw %>%
  select(any_of(c("taxon", "lfc_tetw_RF_preds_std", "q_tetw_RF_preds_std", "passed_ss_tetw_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "tetW exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )


# Read in livestock exposure (modelled) ancom results
ancom_op_cp_genus_meca <- read_csv("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ancom_op_cp_genus_meca.csv")
colnames(ancom_op_cp_genus_meca) <- trimws(colnames(ancom_op_cp_genus_meca))
ancom_op_cp_genus_meca <- as.data.frame(ancom_op_cp_genus_meca)

# mecA
ancom_op_cp_genus_meca <- ancom_op_cp_genus_meca %>%
  select(any_of(c("taxon", "lfc_meca_RF_preds_std", "q_meca_RF_preds_std", "passed_ss_meca_RF_preds_std"))) %>%
  rename_with(~ c("feature", "coef", "q_value", "passed_ss"), 
              .cols = everything()) %>%
  mutate(
    metadata = "mecA exposure",
    label = ifelse(q_value < 0.1, "*", "")
  )



# Bind all exposure dataframes into one dataframe
combined_ancom_op_cp_genus <- bind_rows(
  ancom_op_cp_genus_goat,
  ancom_op_cp_genus_cattle,
  ancom_op_cp_genus_pigs,
  ancom_op_cp_genus_poultry,
  ancom_op_cp_genus_endotoxin, 
  ancom_op_cp_genus_ecoli,
  ancom_op_cp_genus_staph,
  ancom_op_cp_genus_tetw,
  ancom_op_cp_genus_meca
)


combined_ancom_op_cp_genus

# Prepare data for the heatmap
combined_ancom_op_cp_genus <- combined_ancom_op_cp_genus %>%
  select(feature, metadata, coef, label, passed_ss) %>%
  group_by(feature, metadata) 

# Define the custom order of metadata
custom_metadata_order <- c("Endotoxin exposure", "E. coli exposure", "Staph exposure", "tetW exposure", "mecA exposure", "Pigs", "Goats", "Chickens", "Cattle")

# Reorder metadata variables 
combined_ancom_op_cp_genus$metadata <- factor(combined_ancom_op_cp_genus$metadata, levels = custom_metadata_order)

# Modify label colors based on passed_ss
combined_ancom_op_cp_genus <- combined_ancom_op_cp_genus %>%
  mutate(label_color = ifelse(passed_ss == TRUE, "black", "white"))

# Create the heatmap
ancom_heatmap_cp_op_genus <- ggplot(combined_ancom_op_cp_genus, aes(x = metadata, y = feature, fill = coef)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Coefficient"
  ) +
  geom_text(aes(label = label, color = label_color), size = 5) +  # Add labels with color based on passed_ss
  scale_color_identity() +  # Use actual color values for the labels
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin=unit(c(1,1,1,2), 'cm'))

# Save the heatmap
ggsave("../Output/DA_analysis/ANCOM-BC/CP_OP/Genus_level/ANCOM_heatmap_cp_op_genus.svg", 
       ancom_heatmap_cp_op_genus, width = 6, height = 7)
```



## CP and GF 
### ASV level 
#### NP 
```{r}
# Add stars inside the cells based on significance levels (qval thresholds) and passed_ss_populationGF
np_cp_gf_heatmap_data_ancom <- ancom_np_cp_gf_asv_res %>%
  mutate(
    label = case_when(
      q_populationGF < 0.1 ~ "*",    # Stars for qval < 0.1
      TRUE ~ NA_character_           # No stars if qval >= 0.25
    ),
    star_color = ifelse(passed_ss_populationGF, "black", "white")  # Set star color based on passed_ss_populationGF
  )

# Create the heatmap with stars inside the cells
heatmap_cp_gf_np_asvlevel_ancom_stars <- ggplot(np_cp_gf_heatmap_data_ancom, aes(x = "Population", y = taxon, fill = lfc_populationGF)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Log Fold Change"
  ) +
  # Add stars as labels inside the cells, with different colors
  geom_text(aes(label = label, color = star_color), size = 5) +  # Adjust size as needed
  scale_color_identity() +  # Use star_color directly without mapping
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin = unit(c(1, 1, 1, 2), 'cm'))

# Display the heatmap
print(heatmap_cp_gf_np_asvlevel_ancom_stars)

# Save the heatmap to a file
ggsave("../Output/DA_analysis/ANCOM-BC/CP_GF_NP/ASV_level/Heatmap_ANCOMBC2_NP.svg", 
       heatmap_cp_gf_np_asvlevel_ancom_stars, width = 6, height = 8)



```
#### OP
```{r}
# Add stars inside the cells based on significance levels (qval thresholds) and passed_ss_populationGF
op_cp_gf_heatmap_data_ancom <- ancom_op_primary %>%
  mutate(
    label = case_when(
      q_populationGF < 0.1 ~ "*",    # Stars for qval < 0.1
      TRUE ~ NA_character_           # No stars if qval >= 0.25
    ),
    star_color = ifelse(passed_ss_populationGF, "black", "white")  # Set star color based on passed_ss_populationGF
  )


# Create the heatmap with stars inside the cells
heatmap_cp_gf_op_asvlevel_ancom_stars <- ggplot(op_cp_gf_heatmap_data_ancom, aes(x = "Population", y = taxon, fill = lfc_populationGF)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Log Fold Change"
  ) +
  # Add stars as labels inside the cells, with different colors
  geom_text(aes(label = label, color = star_color), size = 5) +  # Adjust size as needed
  scale_color_identity() +  # Use star_color directly without mapping
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin = unit(c(1, 1, 1, 2), 'cm'))

# Display the heatmap
print(heatmap_cp_gf_op_asvlevel_ancom_stars)

# Save the heatmap to a file
ggsave("../Output/DA_analysis/ANCOM-BC/CP_GF_OP/ASV_level/Heatmap_ANCOMBC2_OP.svg", 
       heatmap_cp_gf_op_asvlevel_ancom_stars, width = 6, height = 8)

```

### Genus level 
#### NP 
```{r}
ancom_np_cp_gf_genus_res <- read.csv("../Output/DA_analysis/ANCOM-BC/CP_GF_NP/Genus_level/ancom_np_cp_gf_genus.csv")
# Add stars inside the cells based on significance levels (qval thresholds) and passed_ss_populationGF
np_cp_gf_heatmap_data_ancom_genus <- ancom_np_cp_gf_genus_res %>%
  mutate(
    label = case_when(
      q_populationGF < 0.1 ~ "*",    # Stars for qval < 0.1
      TRUE ~ NA_character_           # No stars if qval >= 0.25
    ),
    star_color = ifelse(passed_ss_populationGF, "black", "white")  # Set star color based on passed_ss_populationGF
  )


# Create the heatmap with stars inside the cells
heatmap_cp_gf_np_genuslevel_ancom_stars <- ggplot(np_cp_gf_heatmap_data_ancom_genus, aes(x = "Population", y = taxon, fill = lfc_populationGF)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Log Fold Change"
  ) +
  # Add stars as labels inside the cells, with different colors
  geom_text(aes(label = label, color = star_color), size = 5) +  # Adjust size as needed
  scale_color_identity() +  # Use star_color directly without mapping
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin = unit(c(1, 1, 1, 2), 'cm'))

# Display the heatmap
print(heatmap_cp_gf_np_genuslevel_ancom_stars)

# Save the heatmap to a file
ggsave("../Output/DA_analysis/ANCOM-BC/CP_GF_NP/Genus_level/Heatmap_ANCOMBC2_NP.svg", 
       heatmap_cp_gf_np_genuslevel_ancom_stars, width = 6, height = 8)



```
#### OP
```{r}
ancom_op_cp_gf_genus_res <- read.csv("../Output/DA_analysis/ANCOM-BC/CP_GF_OP/Genus_level/ancom_op_cp_gf_genus_res.csv")
# Add stars inside the cells based on significance levels (qval thresholds) and passed_ss_populationGF
op_cp_gf_heatmap_data_ancom_genus <- ancom_op_cp_gf_genus_res %>%
  mutate(
    label = case_when(
      q_populationGF < 0.1 ~ "*",    # Stars for qval < 0.1
      TRUE ~ NA_character_           # No stars if qval >= 0.25
    ),
    star_color = ifelse(passed_ss_populationGF, "black", "white")  # Set star color based on passed_ss_populationGF
  )


# Create the heatmap with stars inside the cells
heatmap_cp_gf_op_genuslevel_ancom_stars <- ggplot(op_cp_gf_heatmap_data_ancom_genus, aes(x = "Population", y = taxon, fill = lfc_populationGF)) +
  geom_tile(color = "white") +  # Create the heatmap tiles
  scale_fill_gradient2(
    low = carto_pal(7, "Geyser")[1], 
    mid = carto_pal(7, "Geyser")[4], 
    high = carto_pal(7, "Geyser")[7], 
    midpoint = 0,
    name = "Log Fold Change"
  ) +
  # Add stars as labels inside the cells, with different colors
  geom_text(aes(label = label, color = star_color), size = 5) +  # Adjust size as needed
  scale_color_identity() +  # Use star_color directly without mapping
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels
    axis.title.x = element_blank(),  # Remove x-axis title
    axis.title.y = element_blank(),   # Remove y-axis title
    plot.title = element_text(hjust = 0.5)  # Center the title
  ) +
  theme(plot.margin = unit(c(1, 1, 1, 2), 'cm'))

# Display the heatmap
print(heatmap_cp_gf_op_genuslevel_ancom_stars)

# Save the heatmap to a file
ggsave("../Output/DA_analysis/ANCOM-BC/CP_GF_OP/Genus_level/Heatmap_ANCOMBC2_op.svg", 
       heatmap_cp_gf_op_genuslevel_ancom_stars, width = 6, height = 8)



```



# MaAsLin2 combined heatmaps
## Occupational exposure
```{r}
# Combine the two heatmaps side by side
combined_heatmap_maaslin <- heatmap_cp_gf_np_asvlevel + heatmap_cp_gf_op_asvlevel +
                    plot_layout(ncol = 2)  # Arrange them in 2 columns

# Add a title for the combined plot
combined_heatmap_maaslin <- combined_heatmap_maaslin + 
  plot_annotation(title = "Heatmaps of MaAsLin Results for OP and NP Datasets - Goat farmers vs residents") & 
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))

# Display the combined heatmap
print(combined_heatmap_maaslin)

# Save the combined heatmap
ggsave("../Output/DA_analysis/MaAsLin/Combined_Heatmap_MaAsLin.svg", 
       combined_heatmap_maaslin, width = 10, height = 8)

```
## Environmental exposure
```{r}
# Combine the two heatmaps side by side
combined_hm_maaslin_residential <- heatmap_cp_np_asvlevel + heatmap_cp_op_asvlevel +heatmap_cp_np_genuslevel + heatmap_cp_op_genuslevel+
                    plot_layout(ncol = 2, nrow = 2)  # Arrange them in 2 columns

# Add a title for the combined plot
combined_hm_maaslin_residential <- combined_hm_maaslin_residential + 
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))

# Display the combined heatmap
print(combined_hm_maaslin_residential)

# Save the combined heatmap
ggsave("../Output/DA_analysis/MaAsLin/Heatmap_maaslin_NP_OP_ASV.svg", 
       combined_hm_maaslin_residential, width = 12, height = 12)

```

# ANCOM-BC2 combined heatmaps
## Occupational exposure
```{r}
# NP heatmaps on asv and genus level
ancombc_hm_cp_np_op_combined_occu <- heatmap_cp_gf_np_asvlevel_ancom_stars + heatmap_cp_gf_op_asvlevel_ancom_stars + heatmap_cp_gf_np_genuslevel_ancom_stars +heatmap_cp_gf_op_genuslevel_ancom_stars + 
                    plot_layout(ncol = 2, nrow = 2 )  # Arrange them in 2 columns

ancombc_hm_cp_np_combined_occu <- ancombc_hm_cp_np_combined_occu + 
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))

print(ancombc_hm_cp_np_combined_occu)

ggsave("../Output/DA_analysis/ANCOM-BC/ancombc_hm_cp_np_combined.svg", 
       ancombc_hm_cp_np_combined_occu, width = 11, height = 12)
```

## Environmental exposure 
```{r}
# Combine the two heatmaps side by side
combined_hm_maaslin_residential <- heatmap_cp_np_asv + heatmap_cp_op_asv +heatmap_cp_np_genus + heatmap_cp_op_genus+
                    plot_layout(ncol = 2, nrow = 2)  # Arrange them in 2 columns

# Add a title for the combined plot
combined_hm_maaslin_residential <- combined_hm_maaslin_residential + 
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))

# Display the combined heatmap
print(combined_hm_maaslin_residential)

# Save the combined heatmap
ggsave("../Output/DA_analysis/ANCOM-BC/Heatmap_maaslin_NP_OP_ASV.svg", 
       combined_hm_maaslin_residential, width = 12, height = 12)
```

# Combining MaAsLin and ANCOM-BC 
```{r}
maaslin_ancom_hm_occu <-  heatmap_cp_gf_np_asvlevel + heatmap_cp_gf_np_asvlevel_ancom_stars + heatmap_cp_gf_op_asvlevel + heatmap_cp_gf_op_asvlevel_ancom_stars + heatmap_cp_gf_np_genuslevel + heatmap_cp_gf_np_genuslevel_ancom_stars + heatmap_cp_gf_op_genuslevel + heatmap_cp_gf_op_genuslevel_ancom_stars + 
                    plot_layout(ncol = 4, nrow = 2 )  # Arrange them in 2 columns

maaslin_ancom_hm_occu <- maaslin_ancom_hm_occu + 
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))

ggsave("../Output/DA_analysis/Heatmap_maaslin_ancom_NP_OP_ASV_genus.svg", 
       maaslin_ancom_hm_occu, width = 20, height = 12)

```


# Reporting for article 
```{r}
# MaAsLin2 tends to be more accurate according to the abudance plots
# Both ANCOM-BC2 and MaAsLin2 do a log transformation of the data.
# Since this study is more explorative, MaAsLin resutls will be reported in the main text 
# ANCOM-BC2 results to go in the supplement and referred to 
```


# Check some sequences of interest using BLAST
```{r}
# Make a dataframe of all sequences in the microbiome
seq <- refseq(ps) %>%
  as.data.frame()

# ASVs of interest
which(rownames(seq) == "Moraxella_5") # 5
which(rownames(seq) == "Staphylococcus_2") # 2
which(rownames(seq) == "Corynebacterium_7") # 7
which(rownames(seq) == "Sphingomonas_175") # 131
which(rownames(seq) == "Corynebacterium_stationis_251") #
which(rownames(seq) == "Veillonella_43") # 35
which(rownames(seq) == "Gemella_27") # 24
which(rownames(seq) == "Prevotella_63") # 52
which(rownames(seq) == "Staphylococcaceae_295") # 228
which(rownames(seq) == "Staphylococcus_256") # 199
which(rownames(seq) == "Streptococcus_8") # 8
which(rownames(seq) == "Streptococcus_4") # 4
which(rownames(seq) == "Streptococcus_23") # 21
which(rownames(seq) == "Staphylococcus_217") # 165



tax_table_ps <- tax_table(ps)
print(tax_table_ps["Moraxella_5", ])
print(tax_table_ps["Staphylococcaceae_295", ])
print(tax_table_ps["Staphylococcus_217", ])


# Check the sequences for all of these ASVs
print(seq[5, ]) # Either Moraxella catarrhalis or Moraxella nonliquefaciens
print(seq[2, ])
print(seq[7, ])
print(seq[194, ])
print(seq[35, ])
print(seq[24, ])
print(seq[52, ])
print(seq[228, ]) 
print(seq[199, ])
print(seq[8, ])
print(seq[4, ])
print(seq[21, ])
print(seq[165, ])



tax_table(ps)[7,]
tax_table(ps)[35,]
tax_table(ps)[24,]
tax_table(ps)[52,]
tax_table(ps)[228,]
tax_table(ps)[199,]



# Extract the tax_table from the phyloseq object
tax_table_data <- tax_table(ps)
tax_table_data[1:20,]
# View the filtered data
corynebacterium_data

```
